---
title: "Deeper Sequencing Analysis"
author: "Madison Hobbs"
date: "1/10/2018"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE, echo = FALSE, warning = FALSE, message = FALSE)
```

```{r}
library(tidyverse)
library(stringr)
library(DESeq2)
library(cluster)
library(mclust)

allCounts <- read.csv("DMS2670_LB_NC_000913_deeper.tsv", header = T, sep = "\t")
```

```{r, warning=FALSE, echo=FALSE}
bnum = "b[0-9]{4}"
allCounts$GeneidBackup = allCounts$Geneid
allCounts <- allCounts %>% separate(GeneidBackup, c("feature", "rest"), sep="[:]")

allCounts <- allCounts %>% select(Geneid, feature,                                           
         "0.00_A" = A0,
         "0.00_B" = B0,
         "0.00_C" = C0,
         "0.35_A" = A10..5,
         "0.35_B" = B10..5,
         "0.35_C" = C10..5,
        "11.59_A" = A5x10.5,
         "11.59_B" = B5x10.5,
        "11.59_C" = C5x10.5,
        "20.40_A" = A10..4,
        "20.40_B" = B10..4,
        "20.40_C" = C10..4,
        "48.37_A" = A10..3,
        "48.37_B" = B10..3,
        "48.37_C" = C10..3,
       "100.00_A" = A2537,
       "100.00_B" = B2537,
       "100.00_C" = C2537,
       "129.96_A" = A2x10..3,
       "129.96_B" = B2x10..3,
       "129.96_C" = C2x10..3,
       "190.38_A" = A5x10..3,
       "190.38_B" = B5x10..3,
       "190.38_C" = C5x10..3)

# IGR's separate: 
# do start.bnum end.bnum start.genename end.genename
# left join igrs to allCounts
genename = ",[a-z]{3}[A-Z,]."
rna.name = ",rna[0-9].."
igr <- allCounts %>% filter(feature %in% c("IGR", "AS_IGR"))
igr$GeneidBackup = igr$Geneid
igr <- igr %>% separate(GeneidBackup, c("Geneid1", "Geneid2"), sep = "[/]")
igr$feature1 <- separate(igr, Geneid1, c("feature1", "rest"), sep = "[,]")$feature1
igr$feature1 <- separate(igr, feature1, c("rest", "feature1"), sep = "[()]")$feature1
igr$feature2 <- separate(igr, Geneid2, c("feature2", "rest"), sep = "[,]")$feature2
igr$start.gene <- case_when(
    igr$feature1 == "CDS" ~ str_extract(igr$Geneid1, genename),
    TRUE ~ str_extract(igr$Geneid1, rna.name))
igr$end.gene <- case_when(
    igr$feature2 == "CDS" ~ str_extract(igr$Geneid2, genename),
    TRUE ~ str_extract(igr$Geneid2, rna.name))
igr$start.bnum <- case_when(
    igr$feature1 == "CDS" ~ str_extract(igr$Geneid1, bnum),
    TRUE ~ "none")
igr$end.bnum <- case_when(
    igr$feature2 == "CDS" ~ str_extract(igr$Geneid2, bnum),
    TRUE ~ "none")
igr <- igr %>% separate(start.gene, into = c("comma", "start.gene"), sep = "[,]") %>% select(-comma) %>% separate(end.gene, into = c("comma", "end.gene"), sep = "[,]") %>% select(-comma)
allCounts <- full_join(igr, allCounts)
```


```{r, echo=F, warning = F, message = F}
# CDS
# have bnum and genename columns
# left join to allCounts
genename = ":[a-z]{3}.."
cds <- allCounts %>% filter(feature %in% c("AS_CDS", "CDS")) 
cds$genename <- str_extract(cds$Geneid, genename)
cds$bnum <- str_extract(cds$Geneid, bnum)
cds <- cds %>% separate(genename, into = c("colon", "genename"), sep = ":") %>%
  select(-colon)
allCounts <- full_join(allCounts, cds)
```


```{r, echo=F, warning = F, message = F}
#ncRNA
#ncRNA doesn't have bnums, but id's which we'll put in the genename column
rna.name = ":rna[0-9].."
rna <- allCounts %>% filter(feature %in% c("ncRNA", "AS_ncRNA"))
rna$genename <- str_extract(rna$Geneid, rna.name)
rna <- rna %>% separate(genename, into = c("colon", "genename"), sep = ":") %>%
  select(-colon)
allCounts <- full_join(allCounts, rna)

#rRNA
rRNA <- allCounts %>% filter(feature %in% c("rRNA", "AS_rRNA"))
rRNA$genename <- str_extract(rRNA$Geneid, rna.name)
rRNA <- rRNA %>% separate(genename, into = c("colon", "genename"), sep = ":") %>%
  select(-colon)
allCounts <- full_join(allCounts, rRNA)

#tRNA
tRNA <- allCounts %>% filter(feature %in% c("tRNA", "AS_tRNA"))
tRNA$genename <- str_extract(tRNA$Geneid, rna.name)
tRNA <- tRNA %>% separate(genename, into = c("colon", "genename"), sep = ":") %>%
  select(-colon)
allCounts <- full_join(tRNA, allCounts)

# remove the NA rows we just created by full_joining while adding the ncRNA, rRNA, tRNA genenames
allCounts <- filter(allCounts, feature %in% c("IGR", "AS_IGR") | genename != "NA")

# make tidy data
countsTable.all.tidy <- allCounts %>%
  gather(cond.samps, rawCount, -Geneid, -feature, -genename, -bnum, -Geneid1, -Geneid2, -feature1, -feature2, -start.gene, -end.gene, -start.bnum, - end.bnum)
```

Note: for ease, "deep" data refers to the data just sequenced more deeply while "shallow" data refers to the data we worked on this summer.

The following document compares some diagnostics between the shallow data and deep data, then visualizes the gene expression patterns of both datasets in a clustering setting. 

# Total Counts and Size Factors

### Shallow Data

```{r}
#save deep allCounts
allCounts.deep <- allCounts
countsTable.all.tidy.deep <- countsTable.all.tidy

allCounts <- read.csv("DMS2670_LB_NC_000913.tsv", header = T, sep = "\t") 
allCounts <- allCounts %>%
  group_by(Geneid) %>%
  mutate(genename = ifelse(startsWith(as.character(Geneid), "IGR") | startsWith(as.character(Geneid), "AS_IGR"), paste(unlist(strsplit(as.character(Geneid), "[,]"))[3], "-", unlist(strsplit(as.character(Geneid), "[,]"))[7]), unlist(strsplit(as.character(Geneid), "[:]"))[3]),
         code = unlist(strsplit(as.character(Geneid), "[:]"))[1],
         geneLabel = paste(code, genename)) %>%
  ungroup() %>%
  select(Geneid, geneLabel, code, genename, 
         "0.00_A" = A0, "0.00_B" = B0, "0.00_C" = C0, "0.35_A" = A10..5, "0.35_B" = B10..5, "0.35_C" = C10..5, "11.59_A" = A5x10.5, "11.59_B" = B5x10.5, "11.59_C" = C5x10.5, "20.40_A" = A10..4, "20.40_B" = B10..4, "20.40_C" = C10..4, "48.37_A" = A10..3, "48.37_B" = B10..3, "48.37_C" = C10..3, "129.96_A" = A2x10..3, "129.96_B" = B2x10..3, "129.96_C" = C2x10..3, "190.38_A" = A5x10..3, "190.38_B" = B5x10..3, "190.38_C" = C5x10..3) 
# note: did not the 100.00 samples in the table

# tidy data
countsTable.all.tidy <- allCounts %>%
  gather(cond.samp, rawCount, -Geneid, -geneLabel, -code, -genename) %>%
  mutate(cond.samps = cond.samp) %>%
  separate(cond.samp, c("condition", "sample"), sep = "_")

totalCounts.Ecoli <- allCounts %>% select(`0.00_A`, `0.00_B`, `0.00_C`, 
         `0.35_A`, `0.35_B`, `0.35_C`, 
         `11.59_A`, `11.59_B`, `11.59_C`, 
         `20.40_A`, `20.40_B`, `20.40_C`, 
         `48.37_A`, `48.37_B`, `48.37_C`, 
         `129.96_A`, `129.96_B`, `129.96_C`, 
         `190.38_A`, `190.38_B`, `190.38_C`) %>% summarize_each(funs(sum)) %>%
  gather(cond.samps, total.count)
ggplot(data = totalCounts.Ecoli, aes(x = cond.samps, y=total.count)) + geom_bar(stat = "identity") + ylab("Total Count") + theme(axis.text.x = element_text(angle = 60, hjust=1)) + xlab("Sample") + ggtitle("E. coli Total Counts by Sample")

# use DESeq2 to get size factors counts across all samples, all gene features
countsTable.all <- allCounts %>%
  select(`0.00_A`, `0.00_B`, `0.00_C`, 
         `0.35_A`, `0.35_B`, `0.35_C`, 
         `11.59_A`, `11.59_B`, `11.59_C`, 
         `20.40_A`, `20.40_B`, `20.40_C`, 
         `48.37_A`, `48.37_B`, `48.37_C`, 
         `129.96_A`, `129.96_B`, `129.96_C`, 
         `190.38_A`, `190.38_B`, `190.38_C`)
# define conditions
condition.all <- as.factor(c("0.00", "0.00", "0.00", "0.35", "0.35", "0.35", "11.59", "11.59", "11.59", "20.40",  "20.40",  "20.40", "48.37", "48.37", "48.37", "129.96", "129.96", "129.96", "190.38", "190.38", "190.38"))
coldata.all <- as.data.frame(row.names =  colnames(countsTable.all), condition.all)
cds.all <- DESeqDataSetFromMatrix(countData = countsTable.all, colData = coldata.all, design = ~condition.all)
cds.all <- DESeq(cds.all)
countsTable.all.normalized <- as.data.frame(counts(cds.all, normalized = TRUE)) %>%
  mutate(Geneid = allCounts$Geneid, code = allCounts$code, genename = allCounts$genename)
# tidy countsTable normalized data
countsTable.all.normalized.tidy <-  countsTable.all.normalized %>%
  gather(cond.samp, normCount, -Geneid, -code, -genename) %>%
  mutate(cond.samps = cond.samp) %>%
  separate(cond.samp, c("condition", "sample"), sep = "_") %>%
  mutate(condition = as.numeric(condition))

#extract size factors
allWaySizeFactors <- as.data.frame(sizeFactors(cds.all)) 
cond.samp = rownames(allWaySizeFactors)
allWaySizeFactors <- allWaySizeFactors %>%
  mutate(cond.samp = cond.samp)
seqDepth <- as.data.frame(colSums(countsTable.all)) %>% mutate (cond.samp = cond.samp)

seqDepth.sizeFactors <- inner_join(allWaySizeFactors, seqDepth, by = "cond.samp") %>%
  mutate(sizeFactors.allRegions = sizeFactors(cds.all), 
         sequencingDepth.allRegions = colSums(countsTable.all)) %>%
  select(cond.samp, sizeFactors.allRegions, sequencingDepth.allRegions) %>%
  mutate(cond.samps = cond.samp) %>%
  separate(cond.samp, c("condition", "sample"), sep = "_") %>%
  mutate(condition = as.numeric(condition))

summary.all <- countsTable.all.tidy %>% group_by(cond.samps) %>% summarise(min = min(rawCount), Q1 = quantile(rawCount, .25), median = median(rawCount), Q3 = quantile(rawCount, .75), max = max(rawCount)) %>%
  filter(cond.samps %in% c("100.00_A", "100.00_B", "100.00_C") == FALSE)
summary.all <- inner_join(summary.all, seqDepth.sizeFactors, by = "cond.samps") %>% arrange(as.numeric(condition))
summary.all %>% select(condition, sample, median, Q3, sizeFactor = sizeFactors.allRegions, totalCount = sequencingDepth.allRegions)

# save the shallow tidy table for boxplot later
shallow.countsTable.all.tidy <- countsTable.all.tidy

#replace allCounts with the deep counts
allCounts <- allCounts.deep
```


### Deep Data

```{r}
totalCounts.Ecoli <- allCounts %>% select(`0.00_A`, `0.00_B`, `0.00_C`, 
         `0.35_A`, `0.35_B`, `0.35_C`, 
         `11.59_A`, `11.59_B`, `11.59_C`, 
         `20.40_A`, `20.40_B`, `20.40_C`, 
         `48.37_A`, `48.37_B`, `48.37_C`, 
         `129.96_A`, `129.96_B`, `129.96_C`, 
         `190.38_A`, `190.38_B`, `190.38_C`) %>% summarize_each(funs(sum)) %>%
  gather(cond.samps, total.count)

ggplot(data = totalCounts.Ecoli, aes(x = cond.samps, y=total.count)) + geom_bar(stat = "identity") + ylab("Total Count") + theme(axis.text.x = element_text(angle = 60, hjust=1)) + xlab("Sample") + ggtitle("E. coli Total Counts by Sample")
```


```{r}
# use DESeq2 to get size factors counts across all samples, all gene features
countsTable.all <- allCounts %>%
  select(`0.00_A`, `0.00_B`, `0.00_C`, 
         `0.35_A`, `0.35_B`, `0.35_C`, 
         `11.59_A`, `11.59_B`, `11.59_C`, 
         `20.40_A`, `20.40_B`, `20.40_C`, 
         `48.37_A`, `48.37_B`, `48.37_C`, 
         `129.96_A`, `129.96_B`, `129.96_C`, 
         `190.38_A`, `190.38_B`, `190.38_C`)
# define conditions
condition.all <- as.factor(c("0.00", "0.00", "0.00", "0.35", "0.35", "0.35", "11.59", "11.59", "11.59", "20.40",  "20.40",  "20.40", "48.37", "48.37", "48.37", "129.96", "129.96", "129.96", "190.38", "190.38", "190.38"))
coldata.all <- as.data.frame(row.names =  colnames(countsTable.all), condition.all)
cds.all <- DESeqDataSetFromMatrix(countData = countsTable.all, colData = coldata.all, design = ~condition.all)
cds.all <- DESeq(cds.all)
countsTable.all.normalized <- as.data.frame(counts(cds.all, normalized = TRUE)) %>%
  mutate(Geneid = allCounts$Geneid, feature = allCounts$feature, genename = allCounts$genename)
# tidy countsTable normalized data
countsTable.all.normalized.tidy <-  countsTable.all.normalized %>%
  gather(cond.samp, normCount, -Geneid, -feature, -genename) %>%
  mutate(cond.samps = cond.samp) %>%
  separate(cond.samp, c("condition", "sample"), sep = "_") %>%
  mutate(condition = as.numeric(condition))

#extract size factors
allWaySizeFactors <- as.data.frame(sizeFactors(cds.all)) 
cond.samp = rownames(allWaySizeFactors)
allWaySizeFactors <- allWaySizeFactors %>%
  mutate(cond.samp = cond.samp)
seqDepth <- as.data.frame(colSums(countsTable.all)) %>% mutate (cond.samp = cond.samp)

seqDepth.sizeFactors <- inner_join(allWaySizeFactors, seqDepth, by = "cond.samp") %>%
  mutate(sizeFactors.allRegions = sizeFactors(cds.all), 
         sequencingDepth.allRegions = colSums(countsTable.all)) %>%
  select(cond.samp, sizeFactors.allRegions, sequencingDepth.allRegions) %>%
  mutate(cond.samps = cond.samp) %>%
  separate(cond.samp, c("condition", "sample"), sep = "_") %>%
  mutate(condition = as.numeric(condition))

summary.all <- countsTable.all.tidy %>% group_by(cond.samps) %>% summarise(min = min(rawCount), Q1 = quantile(rawCount, .25), median = median(rawCount), Q3 = quantile(rawCount, .75), max = max(rawCount)) %>%
  filter(cond.samps %in% c("100.00_A", "100.00_B", "100.00_C") == FALSE)
summary.all <- inner_join(summary.all, seqDepth.sizeFactors, by = "cond.samps") %>% arrange(as.numeric(condition))
summary.all %>% select(condition, sample, median, Q3, sizeFactor = sizeFactors.allRegions, totalCount = sequencingDepth.allRegions)
```


We see, unsurprisingly, that the total count per sample of the deeper sequenced data. is larger than that of the shallower sequenced data. However, the size facotors are similar because the distribution of total counts across samples is similar for both depths of sequencing. 

### Shallow Data

```{r, echo=F, fig.height=4}
ggplot(data = shallow.countsTable.all.tidy, aes(x = cond.samps, y = rawCount)) + geom_boxplot() + theme(axis.text.x = element_text(angle = 60, hjust=1)) + xlab("Condition_Sample") + ylab("Read Count") + ggtitle("E. coli Read Count Interquartile Range")

ggplot(data = shallow.countsTable.all.tidy, aes(x = cond.samps, y = rawCount)) + geom_boxplot(outlier.color = "NA", coef = 0) + coord_cartesian(ylim = c(0,60))  + theme(axis.text.x = element_text(angle = 60, hjust=1)) + xlab("Condition_Sample") + ylab("Read Count") + ggtitle("E. coli Read Count Interquartile Range")
```

### Deep Data

```{r, echo=F, fig.height=4}
countsTable.all.tidy <- countsTable.all.tidy.deep
ggplot(data = countsTable.all.tidy, aes(x = cond.samps, y = rawCount)) + geom_boxplot() + theme(axis.text.x = element_text(angle = 60, hjust=1)) + xlab("Condition_Sample") + ylab("Read Count") + ggtitle("E. coli Read Count Interquartile Range")

ggplot(data = countsTable.all.tidy, aes(x = cond.samps, y = rawCount)) + geom_boxplot(outlier.color = "NA", coef = 0) + coord_cartesian(ylim = c(0,300))  + theme(axis.text.x = element_text(angle = 60, hjust=1)) + xlab("Condition_Sample") + ylab("Read Count") + ggtitle("E. coli Read Count Interquartile Range")
```

Note that the y axes on both graphs have different scales, but the distribution of readcounts per sample look similar across the two sequencing depths. the samples with majority lower read count genes in the shallow data still display majority lower read count in the deep data. That is, the samples with the lowest Q3 read count in the shallow data are the same as the samples with lowest read count in the deep data: namely, 0.00_B, 0.35_B, and 20.40_A. The rest of the samples in the deep data follow a similar distribution to those in the shallow data. 

This result could suggest that the deep data is confirming what the shallow data saw, but gene by gene plots are needed to see if this is true.

Before, we eliminated genes for which the maximum unnormalized count of any condition was less than 5. In the shallower sequenced data, there were 3,183 such genes. In the deeper data, we have 938 such genes. This makes sense because the deeper sequenced data is more deeply sequenced!

```{r}
countsTable.all.tidy %>% group_by(Geneid) %>% filter(max(rawCount) < 5) %>% dim()/24
# divide by 24 because that's how many samples there are.
```

# Dominating Genes

### Shallow Data

```{r}
sampleMaxes.Ecoli <- shallow.countsTable.all.tidy %>% group_by(cond.samps) %>%
  summarise(maxCount = max(rawCount)) 
MaxGenes.Ecoli <- shallow.countsTable.all.tidy %>% 
  filter(rawCount %in% sampleMaxes.Ecoli$maxCount) %>% select(Geneid, cond.samps)
sampleMaxes.Ecoli <- inner_join(sampleMaxes.Ecoli, MaxGenes.Ecoli, by = "cond.samps")
sampleMaxes.Ecoli <- inner_join(sampleMaxes.Ecoli, totalCounts.Ecoli, by = "cond.samps") %>% mutate(propTotalCount = maxCount/total.count)

tenDominating <- shallow.countsTable.all.tidy %>% arrange(desc(rawCount)) %>% distinct(Geneid, .keep_all = TRUE) %>% select(rawCount, Geneid) %>% head(10)
```

### Deep Data

```{r}
countsTable.all.tidy <- countsTable.all.tidy.deep

sampleMaxes.Ecoli <- countsTable.all.tidy %>% group_by(cond.samps) %>%
  summarise(maxCount = max(rawCount)) 
MaxGenes.Ecoli <- countsTable.all.tidy %>% 
  filter(rawCount %in% sampleMaxes.Ecoli$maxCount) %>% select(Geneid, cond.samps)
sampleMaxes.Ecoli <- inner_join(sampleMaxes.Ecoli, MaxGenes.Ecoli, by = "cond.samps")
sampleMaxes.Ecoli <- inner_join(sampleMaxes.Ecoli, totalCounts.Ecoli, by = "cond.samps") %>% mutate(propTotalCount = maxCount/total.count)

tenDominating <- countsTable.all.tidy %>% arrange(desc(rawCount)) %>% distinct(Geneid, .keep_all = TRUE) %>% select(rawCount, Geneid) %>% head(10)
```

Of the ten most expressed genes from each level of sequencing, eight are the same in both levels of sequencing. These are smpB-intA, rna69, rna106, rmf b0953, rna98, cspE, rna54, and rna71.

```{r}
shallow.dominating <- shallow.countsTable.all.tidy %>% arrange(desc(rawCount)) %>% distinct(Geneid, .keep_all = TRUE) %>% select(rawCount, Geneid) %>% head(1000)

deep.dominating <- countsTable.all.tidy %>% arrange(desc(rawCount)) %>% distinct(Geneid, .keep_all = TRUE) %>% select(rawCount, Geneid) %>% head(1000)

inner_join(shallow.dominating, deep.dominating, by = "Geneid") %>% dim() 
anti_join(shallow.dominating, deep.dominating, by = "Geneid") %>% dim() 
```

In fact, of the 1000 most expressed genes in each level of sequencing, 844/1000 are the same and 156/1000 are different. This shows that the genes which dominated the shallow data tend to be the same genes dominating the deep data. 

# Differential Expression, Gene Expression Shapes, and Clustering

We throw out the samples with much lower median read count (0.00_B, 0.35_B, and 20.40_A) as well as the samples with 100% RpoS (100.00_A, 100.00_B, 100.00_C) because it has a different strain of E. coli than the rest of the samples. 

###Shallow Data

```{r, warning=FALSE, echo=FALSE}
allCounts <- read.csv("DMS2670_LB_NC_000913.tsv", header = T, sep = "\t") 
bnum = "b[0-9]{4}"
allCounts$GeneidBackup = allCounts$Geneid
allCounts <- allCounts %>% separate(GeneidBackup, c("feature", "rest"), sep="[:]")

allCounts <- allCounts %>% select(Geneid, feature,                                          
         "0.00_A" = A0,
         "0.00_B" = B0,
         "0.00_C" = C0,
         "0.35_A" = A10..5,
         "0.35_B" = B10..5,
         "0.35_C" = C10..5,
        "11.59_A" = A5x10.5,
         "11.59_B" = B5x10.5,
        "11.59_C" = C5x10.5,
        "20.40_A" = A10..4,
        "20.40_B" = B10..4,
        "20.40_C" = C10..4,
        "48.37_A" = A10..3,
        "48.37_B" = B10..3,
        "48.37_C" = C10..3,
       "100.00_A" = A2537,
       "100.00_B" = B2537,
       "100.00_C" = C2537,
       "129.96_A" = A2x10..3,
       "129.96_B" = B2x10..3,
       "129.96_C" = C2x10..3,
       "190.38_A" = A5x10..3,
       "190.38_B" = B5x10..3,
       "190.38_C" = C5x10..3)

# IGR's separate: 
# do start.bnum end.bnum start.genename end.genename
# left join igrs to allCounts
genename = ",[a-z]{3}[A-Z,]."
rna.name = ",rna[0-9].."
igr <- allCounts %>% filter(feature %in% c("IGR", "AS_IGR"))
igr$GeneidBackup = igr$Geneid
igr <- igr %>% separate(GeneidBackup, c("Geneid1", "Geneid2"), sep = "[/]")
igr$feature1 <- separate(igr, Geneid1, c("feature1", "rest"), sep = "[,]")$feature1
igr$feature1 <- separate(igr, feature1, c("rest", "feature1"), sep = "[()]")$feature1
igr$feature2 <- separate(igr, Geneid2, c("feature2", "rest"), sep = "[,]")$feature2
igr$start.gene <- case_when(
    igr$feature1 == "CDS" ~ str_extract(igr$Geneid1, genename),
    TRUE ~ str_extract(igr$Geneid1, rna.name))
igr$end.gene <- case_when(
    igr$feature2 == "CDS" ~ str_extract(igr$Geneid2, genename),
    TRUE ~ str_extract(igr$Geneid2, rna.name))
igr$start.bnum <- case_when(
    igr$feature1 == "CDS" ~ str_extract(igr$Geneid1, bnum),
    TRUE ~ "none")
igr$end.bnum <- case_when(
    igr$feature2 == "CDS" ~ str_extract(igr$Geneid2, bnum),
    TRUE ~ "none")
igr <- igr %>% separate(start.gene, into = c("comma", "start.gene"), sep = "[,]") %>% select(-comma) %>% separate(end.gene, into = c("comma", "end.gene"), sep = "[,]") %>% select(-comma)
allCounts <- full_join(igr, allCounts)

# CDS
# have bnum and genename columns
# left join to allCounts
genename = ":[a-z]{3}.."
cds <- allCounts %>% filter(feature %in% c("AS_CDS", "CDS")) 
cds$genename <- str_extract(cds$Geneid, genename)
cds$bnum <- str_extract(cds$Geneid, bnum)
cds <- cds %>% separate(genename, into = c("colon", "genename"), sep = ":") %>%
  select(-colon)
allCounts <- full_join(allCounts, cds)

#ncRNA
#ncRNA doesn't have bnums, but id's which we'll put in the genename column
rna.name = ":rna[0-9].."
rna <- allCounts %>% filter(feature %in% c("ncRNA", "AS_ncRNA"))
rna$genename <- str_extract(rna$Geneid, rna.name)
rna <- rna %>% separate(genename, into = c("colon", "genename"), sep = ":") %>%
  select(-colon)
allCounts <- full_join(allCounts, rna)

#rRNA
rRNA <- allCounts %>% filter(feature %in% c("rRNA", "AS_rRNA"))
rRNA$genename <- str_extract(rRNA$Geneid, rna.name)
rRNA <- rRNA %>% separate(genename, into = c("colon", "genename"), sep = ":") %>%
  select(-colon)
allCounts <- full_join(allCounts, rRNA)

#tRNA
tRNA <- allCounts %>% filter(feature %in% c("tRNA", "AS_tRNA"))
tRNA$genename <- str_extract(tRNA$Geneid, rna.name)
tRNA <- tRNA %>% separate(genename, into = c("colon", "genename"), sep = ":") %>%
  select(-colon)
allCounts <- full_join(tRNA, allCounts)

# remove the NA rows we just created by full_joining while adding the ncRNA, rRNA, tRNA genenames
allCounts <- filter(allCounts, feature %in% c("IGR", "AS_IGR") | genename != "NA")

# make tidy data
countsTable.all.tidy <- allCounts %>%
  gather(cond.samps, rawCount, -Geneid, -feature, -genename, -bnum, -Geneid1, -Geneid2, -feature1, -feature2, -start.gene, -end.gene, -start.bnum, - end.bnum)
```

```{r, echo=F, message=F, warning=F}
# keep only genes whose maximum raw count is greater than 5 and exclude the samples `0.00_B`, `0.35_B`, and `20.40_A` whose Q3's are below 10. We also exclude the samples with condition 100% since these are a different strain of E. Coli
allCounts.tidy.filtered <- countsTable.all.tidy %>% group_by(Geneid) %>%
  filter(max(rawCount) > 5) %>% filter(cond.samps %in% c("0.00_B", "0.35_B", "20.40_A", "100.00_A", "100.00_B", "100.00_C") == FALSE)

# create wide, messy version of this which DESeq2 needs. Filter out the unwanted columns but not the unwanted rows; we're going to need these for normalization
allCounts.filtered <- allCounts %>% select(Geneid, feature, `0.00_A`, `0.00_C`,
         `0.35_A`, `0.35_C`, 
         `11.59_A`, `11.59_B`, `11.59_C`, 
         `20.40_B`, `20.40_C`, 
         `48.37_A`, `48.37_B`, `48.37_C`, 
         `129.96_A`, `129.96_B`, `129.96_C`, 
         `190.38_A`, `190.38_B`, `190.38_C`)

allCounts.filtered.justCounts <- allCounts.filtered %>% select(`0.00_A`, `0.00_C`, 
 `0.35_A`, `0.35_C`,  `11.59_A`, `11.59_B`, `11.59_C`, `20.40_B`, `20.40_C`, 
`48.37_A`, `48.37_B`, `48.37_C`, `129.96_A`, `129.96_B`, `129.96_C`,  
`190.38_A`, `190.38_B`, `190.38_C`)

condition.filtered <- as.factor(c("0.00", "0.00", "0.35", "0.35", "11.59", "11.59", "11.59", "20.40", "20.40", "48.37", "48.37", "48.37", "129.96", "129.96", "129.96", "190.38", "190.38", "190.38"))

coldata.filtered <- as.data.frame(row.names =  colnames(allCounts.filtered.justCounts), condition.filtered)
dds.filtered <- DESeqDataSetFromMatrix(countData = allCounts.filtered.justCounts, colData = coldata.filtered, design = ~condition.filtered)
# estimate size factors, dispersions, etc, and testing differential expression
dds.filtered <- DESeq(dds.filtered)

# get normalized counts
normCounts.filtered <- as.data.frame(counts(dds.filtered, normalized = TRUE)) %>%
  dplyr::rename(norm0.00_A = `0.00_A`, norm0.00_C = `0.00_C`, norm0.35_A = `0.35_A`, norm0.35_C = `0.35_C`, norm11.59_A = `11.59_A`, norm11.59_B = `11.59_B`, norm11.59_C = `11.59_C`, norm20.40_B = `20.40_B`,norm20.40_C = `20.40_C`, norm48.37_A = `48.37_A`,  norm48.37_B = `48.37_B`, norm48.37_C = `48.37_C`, norm129.96_A = `129.96_A`, norm129.96_B = `129.96_B`, norm129.96_C = `129.96_C`, norm190.38_A = `190.38_A`, norm190.38_B = `190.38_B`, norm190.38_C = `190.38_C`)
# append normalized counts to wide, messy allCounts.filtered.
# ALSO FILTER TO REMOVE THE GENES WHOSE MAX RAW COUNT WAS LESS THAN 5
allCounts.filtered <- cbind(allCounts.filtered, normCounts.filtered) %>% 
  filter(Geneid %in% allCounts.tidy.filtered$Geneid) 
  #these are the genes whose max raw count was less than 5

#get tidy normalized counts
normCounts.tidy.filtered <- allCounts.filtered %>% select(Geneid, norm0.00_A, norm0.00_C, norm0.35_A, norm0.35_C, norm11.59_A, norm11.59_B, norm11.59_C, norm20.40_B, norm20.40_C, norm48.37_A,  norm48.37_B, norm48.37_C, norm129.96_A, norm129.96_B, norm129.96_C, norm190.38_A, norm190.38_B, norm190.38_C) %>%
  #change back to original names (this is a bit cumbersome, sorry!)
dplyr::rename(`0.00_A` = norm0.00_A, `0.00_C` = norm0.00_C, 
                `0.35_A` = norm0.35_A, `0.35_C` = norm0.35_C, 
            `11.59_A` = norm11.59_A, `11.59_B` = norm11.59_B, `11.59_C` =  norm11.59_C, `20.40_B` = norm20.40_B, `20.40_C` = norm20.40_C, 
            `48.37_A` = norm48.37_A, `48.37_B` = norm48.37_B, `48.37_C` = norm48.37_C, `129.96_A` = norm129.96_A, `129.96_B` = norm129.96_B, `129.96_C` = norm129.96_C, `190.38_A` = norm190.38_A, `190.38_B` = norm190.38_B, `190.38_C` = norm190.38_C) %>%
  gather(cond.samps, normCount, -Geneid)

allCounts.tidy.filtered <- inner_join(allCounts.tidy.filtered, normCounts.tidy.filtered, by = c("Geneid", "cond.samps"))

#make allCounts.tidy.filtered ready for plotting
allCounts.tidy.filtered <- allCounts.tidy.filtered %>%
  separate(cond.samps, c("RpoS%", "replicate"), sep = "[_]") %>%
  mutate(`RpoS%` = as.numeric(`RpoS%`)) %>%
  group_by(Geneid, `RpoS%`) %>%
  mutate(levelMean = mean(normCount)) %>%
  ungroup() %>%
  group_by(Geneid) %>%
  mutate(scaler = ifelse(levelMean[1] < levelMean[18], levelMean[18], levelMean[1]),
         normCountScaled = normCount/scaler,
         meanScaled = levelMean/scaler) %>%
  ungroup() 
```

```{r, echo=F, message=F, warning=F}
# compare 0 to 190.38
countsTable.filtered.0.190.38 <- allCounts.filtered %>%
  select(`0.00_A`, `0.00_C`, `190.38_A`, `190.38_B`, `190.38_C`)
rownames(countsTable.filtered.0.190.38) <- allCounts.filtered$Geneid
# define conditions
condition.filtered.0.190.38 <- as.factor(c("0.00", "0.00", "190.38", "190.38", "190.38"))
# we only want the relevant column name from countsTable for each condition (we don't want Geneid or genename)
rownames <- as.data.frame(colnames(countsTable.filtered.0.190.38))
coldata.filtered.0.190.38 <- data.frame(row.names = colnames(countsTable.filtered.0.190.38), condition.filtered.0.190.38)
# create the DESeqDataSet. cds stands for "count data set"
dds.filtered.0.190.38 <- DESeqDataSetFromMatrix(countData = countsTable.filtered.0.190.38, colData = coldata.filtered.0.190.38, design = ~condition.filtered.0.190.38)
# estimate size factors, dispersions, etc, and testing differential expression
dds.filtered.0.190.38 <- DESeq(dds.filtered.0.190.38)
#estimating dispersion by treating samples as replicates

#get results
results.filtered.0.190.38 <- results(dds.filtered.0.190.38, alpha = 0.05)
summary(results.filtered.0.190.38)

Geneids <- rownames(countsTable.filtered.0.190.38)
regulated.genes.filtered <- as.data.frame(results.filtered.0.190.38) %>% mutate(Geneid = Geneids, regulation = ifelse(log2FoldChange > 0, "positive", "negative"))
regulated.genes.filtered <- filter(regulated.genes.filtered, padj < 0.05)

#modify allCounts.filtered to include regulation information
allCounts.filtered <- left_join(allCounts.filtered, regulated.genes.filtered, by = "Geneid")

#modify allCounts.tidy.filterd to include regulation information
allCounts.tidy.filtered <- left_join(allCounts.tidy.filtered, regulated.genes.filtered, by = "Geneid")
```

```{r, echo=F}
#gather nice wide data frame of normalized counts to cluster 
countsToCluster <- allCounts.filtered %>% filter(Geneid %in% regulated.genes.filtered$Geneid) %>%
   select(norm0.00_A, norm0.00_C, norm0.35_A, norm0.35_C, norm11.59_A, norm11.59_B, norm11.59_C, norm20.40_B, norm20.40_C, norm48.37_A,  norm48.37_B, norm48.37_C, norm129.96_A, norm129.96_B, norm129.96_C, norm190.38_A, norm190.38_B, norm190.38_C)
#save Geneid's as rownames
rownames(countsToCluster) <- filter(allCounts.filtered, Geneid %in% regulated.genes.filtered$Geneid)$Geneid

#generate dissmilarity matrix based on spearman correlation
dissimilarity.matrix <- 1 - cor(t(countsToCluster), method = "spearman")

pam.clusters <- c()
pam.medoids <- c()
  pam <- pam(dissimilarity.matrix, 6)
  thisCluster <- as.data.frame(pam$clustering)
  pam.medoids <- c(pam.medoids, rownames(pam$medoids))
  
pam.clusters <- thisCluster %>%
  mutate(Geneid = rownames(countsToCluster))
pam.medoids <- data.frame(pam.medoids, 6)
  
allCounts.tidy.filtered <- left_join(allCounts.tidy.filtered, pam.clusters) 
shallow.regulated.tidy <- filter(allCounts.tidy.filtered, regulation != "NA")
```

```{r}
medoids.k6 <- filter(shallow.regulated.tidy, Geneid %in% pam.medoids$pam.medoids)
```

```{r}
#k = 6
#show medoids alone
ggplot(data = medoids.k6, aes(x = `RpoS%`, y = normCountScaled)) + geom_point() + geom_line(data = medoids.k6, aes(y = meanScaled), size = 2) + coord_cartesian(ylim = c(0,2)) + theme(legend.position = "none") + facet_grid(. ~ `pam$clustering`) + ggtitle("PAM medoids, k=6")+ ylab("Gene Expression")
```

```{r}
# k = 6, show genes with medoids
ggplot(data = shallow.regulated.tidy, aes(x = `RpoS%`, y = normCountScaled, col=Geneid)) + geom_point() + geom_line(data = shallow.regulated.tidy, aes(y = meanScaled), linetype = "dashed") + coord_cartesian(ylim = c(0,2)) + theme(legend.position = "none") + facet_grid(. ~ `pam$clustering`) + ggtitle("Gene Expression Clustering, PAM k=6; Medoids Overlaid in black") + ylab("Gene Expression") + geom_line(data = medoids.k6, aes(y = meanScaled), col = "black", size = 2) + theme(plot.title = element_text(face = "bold"))
```


### Deep Data, all samples except 100% RpoS

```{r}
#reset all Counts to be the deep version
allCounts <- allCounts.deep
countsTable.all.tidy <- countsTable.all.tidy.deep

allCounts.tidy.filtered <- countsTable.all.tidy %>% group_by(Geneid) %>%
  filter(max(rawCount) > 5)

allCounts.filtered <- allCounts %>% select(-`100.00_A`, -`100.00_B`, -`100.00_C`)

allCounts.filtered.justCounts <- allCounts.filtered %>% select(-Geneid, -feature, -Geneid1, -Geneid2, -feature1, -feature2, -start.gene, -end.gene, -start.bnum, -end.bnum, -genename, -bnum)

condition.filtered <- as.factor(c("0.00", "0.00", "0.00", "0.35", "0.35", "0.35", "11.59", "11.59", "11.59", "20.40", "20.40", "20.40", "48.37", "48.37", "48.37", "129.96", "129.96", "129.96", "190.38", "190.38", "190.38"))

coldata.filtered <- as.data.frame(row.names =  colnames(allCounts.filtered.justCounts), condition.filtered)

dds.filtered <- DESeqDataSetFromMatrix(countData = allCounts.filtered.justCounts, colData = coldata.filtered, design = ~condition.filtered)

dds.filtered <- DESeq(dds.filtered)

# get normalized counts
normCounts.filtered <- as.data.frame(counts(dds.filtered, normalized = TRUE)) %>%
  dplyr::rename(norm0.00_A = `0.00_A`,norm0.00_B = `0.00_B`, norm0.00_C = `0.00_C`, 
                norm0.35_A = `0.35_A`,  norm0.35_B = `0.35_B`, norm0.35_C = `0.35_C`, 
                norm11.59_A = `11.59_A`, norm11.59_B = `11.59_B`, norm11.59_C = `11.59_C`, 
                norm20.40_A = `20.40_A`, norm20.40_B = `20.40_B`, norm20.40_C = `20.40_C`, 
                norm48.37_A = `48.37_A`,  norm48.37_B = `48.37_B`, norm48.37_C = `48.37_C`,  
                norm129.96_A = `129.96_A`, norm129.96_B = `129.96_B`, norm129.96_C = `129.96_C`, 
                norm190.38_A = `190.38_A`, norm190.38_B = `190.38_B`, norm190.38_C = `190.38_C`)

# append normalized counts to wide, messy allCounts.filtered.
# ALSO FILTER TO REMOVE THE GENES WHOSE MAX RAW COUNT WAS LESS THAN 5
allCounts.filtered <- cbind(allCounts.filtered, normCounts.filtered) %>% 
  filter(Geneid %in% allCounts.tidy.filtered$Geneid) 
  #these are the genes whose max raw count was less than 5

#get tidy normalized counts
normCounts.tidy.filtered <- allCounts.filtered %>% select(Geneid, norm0.00_A, norm0.00_B, norm0.00_C, norm0.35_A,  norm0.35_B, norm0.35_C, norm11.59_A, norm11.59_B, norm11.59_C, norm20.40_A, norm20.40_B, norm20.40_C, norm48.37_A,  norm48.37_B, norm48.37_C, norm129.96_A, norm129.96_B, norm129.96_C, norm190.38_A, norm190.38_B, norm190.38_C) %>%
  #change back to original names (this is a bit cumbersome, sorry!)
dplyr::rename(`0.00_A` = norm0.00_A, `0.00_B` = norm0.00_B, `0.00_C` = norm0.00_C, 
                `0.35_A` = norm0.35_A,`0.35_B` = norm0.35_B, `0.35_C` = norm0.35_C, 
            `11.59_A` = norm11.59_A, `11.59_B` = norm11.59_B, `11.59_C` =  norm11.59_C, `20.40_A` = norm20.40_A, `20.40_B` = norm20.40_B, `20.40_C` = norm20.40_C, 
            `48.37_A` = norm48.37_A, `48.37_B` = norm48.37_B, `48.37_C` = norm48.37_C,
            `129.96_A` = norm129.96_A, `129.96_B` = norm129.96_B, `129.96_C` = norm129.96_C, `190.38_A` = norm190.38_A, `190.38_B` = norm190.38_B, `190.38_C` = norm190.38_C) %>%
  gather(cond.samps, normCount, -Geneid)

allCounts.tidy.filtered <- inner_join(allCounts.tidy.filtered, normCounts.tidy.filtered, by = c("Geneid", "cond.samps"))

allCounts.tidy.filtered <- allCounts.tidy.filtered %>%
  separate(cond.samps, c("RpoS%", "replicate"), sep = "[_]") %>%
  mutate(`RpoS%` = as.numeric(`RpoS%`)) %>%
  group_by(Geneid, `RpoS%`) %>%
  mutate(levelMean = mean(normCount)) %>%
  ungroup() %>%
  group_by(Geneid) %>%
  mutate(scaler = ifelse(levelMean[1] < levelMean[21], levelMean[21], levelMean[1]),
         normCountScaled = normCount/scaler,
         meanScaled = levelMean/scaler) %>%
  ungroup() 
```
.

```{r, echo=F, message=F, warning=F}
# compare 0 to 190.38
countsTable.filtered.0.190.38 <- allCounts.filtered %>%
  select(`0.00_A`, `0.00_B`, `0.00_C`, `190.38_A`, `190.38_B`, `190.38_C`)
rownames(countsTable.filtered.0.190.38) <- allCounts.filtered$Geneid
# define conditions
condition.filtered.0.190.38 <- as.factor(c("0.00", "0.00", "0.00", "190.38", "190.38", "190.38"))
# we only want the relevant column name from countsTable for each condition (we don't want Geneid or genename)
rownames <- as.data.frame(colnames(countsTable.filtered.0.190.38))
coldata.filtered.0.190.38 <- data.frame(row.names = colnames(countsTable.filtered.0.190.38), condition.filtered.0.190.38)
# create the DESeqDataSet. cds stands for "count data set"
dds.filtered.0.190.38 <- DESeqDataSetFromMatrix(countData = countsTable.filtered.0.190.38, colData = coldata.filtered.0.190.38, design = ~condition.filtered.0.190.38)
# estimate size factors, dispersions, etc, and testing differential expression
dds.filtered.0.190.38 <- DESeq(dds.filtered.0.190.38)
#estimating dispersion by treating samples as replicates

#get results
results.filtered.0.190.38 <- results(dds.filtered.0.190.38, alpha = 0.05)
summary(results.filtered.0.190.38)

Geneids <- rownames(countsTable.filtered.0.190.38)
regulated.genes.filtered <- as.data.frame(results.filtered.0.190.38) %>% mutate(Geneid = Geneids, regulation = ifelse(log2FoldChange > 0, "positive", "negative"))
regulated.genes.filtered <- filter(regulated.genes.filtered, padj < 0.05)

#modify allCounts.filtered to include regulation information
allCounts.filtered <- left_join(allCounts.filtered, regulated.genes.filtered, by = "Geneid")

#modify allCounts.tidy.filterd to include regulation information
allCounts.tidy.filtered <- left_join(allCounts.tidy.filtered, regulated.genes.filtered, by = "Geneid")
```

```{r, echo = F}
#gather nice wide data frame of normalized counts to cluster 
countsToCluster <- allCounts.filtered %>% filter(Geneid %in% regulated.genes.filtered$Geneid) %>%
   select(norm0.00_A, norm0.00_B, norm0.00_C, norm0.35_A, norm0.35_B, norm0.35_C, norm11.59_A, norm11.59_B, norm11.59_C, norm20.40_A, norm20.40_B, norm20.40_C, norm48.37_A,  norm48.37_B, norm48.37_C, norm129.96_A, norm129.96_B, norm129.96_C, norm190.38_A, norm190.38_B, norm190.38_C)
#save Geneid's as rownames
rownames(countsToCluster) <- filter(allCounts.filtered, Geneid %in% regulated.genes.filtered$Geneid)$Geneid

#generate dissmilarity matrix based on spearman correlation
dissimilarity.matrix <- 1 - cor(t(countsToCluster), method = "spearman")

pam.clusters <- c()
pam.medoids <- c()
  pam <- pam(dissimilarity.matrix, 6)
  thisCluster <- as.data.frame(pam$clustering)
  pam.medoids <- c(pam.medoids, rownames(pam$medoids))
  
pam.clusters <- thisCluster %>%
  mutate(Geneid = rownames(countsToCluster))
pam.medoids <- data.frame(pam.medoids, 6)
  
allCounts.tidy.filtered <- left_join(allCounts.tidy.filtered, pam.clusters) 
extra.deep.regulated.tidy <- filter(allCounts.tidy.filtered, regulation != "NA")
```

```{r}
extra.medoids.k6 <- filter(extra.deep.regulated.tidy, Geneid %in% pam.medoids$pam.medoids)
```

```{r}
#k = 6
#show medoids alone
ggplot(data = extra.medoids.k6, aes(x = `RpoS%`, y = normCountScaled)) + geom_point() + geom_line(data = extra.medoids.k6, aes(y = meanScaled), size = 2) + coord_cartesian(ylim = c(0,2.3)) + theme(legend.position = "none") + facet_grid(. ~ `pam$clustering`) + ggtitle("PAM medoids, k=6")+ ylab("Gene Expression")
```

```{r}
# k = 6, show genes with medoids
ggplot(data = extra.deep.regulated.tidy, aes(x = `RpoS%`, y = normCountScaled, col=Geneid)) + geom_point() + geom_line(data = extra.deep.regulated.tidy, aes(y = meanScaled), linetype = "dashed") + coord_cartesian(ylim = c(0,2.3)) + theme(legend.position = "none") + facet_grid(. ~ `pam$clustering`) + ggtitle("Gene Expression Clustering, PAM k=6; Medoids Overlaid in black") + ylab("Gene Expression") + geom_line(data = extra.medoids.k6, aes(y = meanScaled), col = "black", size = 2) + theme(plot.title = element_text(face = "bold"))
```

### Deep Data, 0.00_B, 0.35_B, 20.40_A and  100% RpoS removed

We remove the samples with 100% RpoS and 0.00_B, 0.35_B, 20.40_A.

```{r}
#reset all Counts to be the deep version
allCounts <- allCounts.deep
countsTable.all.tidy <- countsTable.all.tidy.deep

allCounts.tidy.filtered <- countsTable.all.tidy %>% group_by(Geneid) %>%
  filter(max(rawCount) > 5)

allCounts.filtered <- allCounts %>% select(-`0.00_B`, -`0.35_B`, -`20.40_A`, -`100.00_A`, -`100.00_B`, -`100.00_C`)

allCounts.filtered.justCounts <- allCounts.filtered %>% select(-Geneid, -feature, -Geneid1, -Geneid2, -feature1, -feature2, -start.gene, -end.gene, -start.bnum, -end.bnum, -genename, -bnum)

condition.filtered <- as.factor(c("0.00", "0.00", "0.35", "0.35", "11.59", "11.59", "11.59", "20.40", "20.40", "48.37", "48.37", "48.37", "129.96", "129.96", "129.96", "190.38", "190.38", "190.38"))

coldata.filtered <- as.data.frame(row.names =  colnames(allCounts.filtered.justCounts), condition.filtered)

dds.filtered <- DESeqDataSetFromMatrix(countData = allCounts.filtered.justCounts, colData = coldata.filtered, design = ~condition.filtered)

dds.filtered <- DESeq(dds.filtered)

# get normalized counts
normCounts.filtered <- as.data.frame(counts(dds.filtered, normalized = TRUE)) %>%
  dplyr::rename(norm0.00_A = `0.00_A`, norm0.00_C = `0.00_C`, norm0.35_A = `0.35_A`, norm0.35_C = `0.35_C`, norm11.59_A = `11.59_A`, norm11.59_B = `11.59_B`, norm11.59_C = `11.59_C`, norm20.40_B = `20.40_B`, norm20.40_C = `20.40_C`, norm48.37_A = `48.37_A`,  norm48.37_B = `48.37_B`, norm48.37_C = `48.37_C`,  norm129.96_A = `129.96_A`, norm129.96_B = `129.96_B`, norm129.96_C = `129.96_C`, norm190.38_A = `190.38_A`, norm190.38_B = `190.38_B`, norm190.38_C = `190.38_C`)

# append normalized counts to wide, messy allCounts.filtered.
# ALSO FILTER TO REMOVE THE GENES WHOSE MAX RAW COUNT WAS LESS THAN 5
allCounts.filtered <- cbind(allCounts.filtered, normCounts.filtered) %>% 
  filter(Geneid %in% allCounts.tidy.filtered$Geneid) 
  #these are the genes whose max raw count was less than 5

#get tidy normalized counts
normCounts.tidy.filtered <- allCounts.filtered %>% select(Geneid, norm0.00_A, norm0.00_C, norm0.35_A, norm0.35_C, norm11.59_A, norm11.59_B, norm11.59_C, norm20.40_B, norm20.40_C, norm48.37_A,  norm48.37_B, norm48.37_C, norm129.96_A, norm129.96_B, norm129.96_C, norm190.38_A, norm190.38_B, norm190.38_C) %>%
  #change back to original names (this is a bit cumbersome, sorry!)
dplyr::rename(`0.00_A` = norm0.00_A, `0.00_C` = norm0.00_C, 
                `0.35_A` = norm0.35_A, `0.35_C` = norm0.35_C, 
            `11.59_A` = norm11.59_A, `11.59_B` = norm11.59_B, `11.59_C` =  norm11.59_C, `20.40_B` = norm20.40_B, `20.40_C` = norm20.40_C, 
            `48.37_A` = norm48.37_A, `48.37_B` = norm48.37_B, `48.37_C` = norm48.37_C,
            `129.96_A` = norm129.96_A, `129.96_B` = norm129.96_B, `129.96_C` = norm129.96_C, `190.38_A` = norm190.38_A, `190.38_B` = norm190.38_B, `190.38_C` = norm190.38_C) %>%
  gather(cond.samps, normCount, -Geneid)

allCounts.tidy.filtered <- inner_join(allCounts.tidy.filtered, normCounts.tidy.filtered, by = c("Geneid", "cond.samps"))

allCounts.tidy.filtered <- allCounts.tidy.filtered %>%
  separate(cond.samps, c("RpoS%", "replicate"), sep = "[_]") %>%
  mutate(`RpoS%` = as.numeric(`RpoS%`)) %>%
  group_by(Geneid, `RpoS%`) %>%
  mutate(levelMean = mean(normCount)) %>%
  ungroup() %>%
  group_by(Geneid) %>%
  mutate(scaler = ifelse(levelMean[1] < levelMean[18], levelMean[18], levelMean[1]),
         normCountScaled = normCount/scaler,
         meanScaled = levelMean/scaler) %>%
  ungroup() 
```
.

```{r, echo=F, message=F, warning=F}
# compare 0 to 190.38
countsTable.filtered.0.190.38 <- allCounts.filtered %>%
  select(`0.00_A`, `0.00_C`, `190.38_A`, `190.38_B`, `190.38_C`)
rownames(countsTable.filtered.0.190.38) <- allCounts.filtered$Geneid
# define conditions
condition.filtered.0.190.38 <- as.factor(c("0.00", "0.00", "190.38", "190.38", "190.38"))
# we only want the relevant column name from countsTable for each condition (we don't want Geneid or genename)
rownames <- as.data.frame(colnames(countsTable.filtered.0.190.38))
coldata.filtered.0.190.38 <- data.frame(row.names = colnames(countsTable.filtered.0.190.38), condition.filtered.0.190.38)
# create the DESeqDataSet. cds stands for "count data set"
dds.filtered.0.190.38 <- DESeqDataSetFromMatrix(countData = countsTable.filtered.0.190.38, colData = coldata.filtered.0.190.38, design = ~condition.filtered.0.190.38)
# estimate size factors, dispersions, etc, and testing differential expression
dds.filtered.0.190.38 <- DESeq(dds.filtered.0.190.38)
#estimating dispersion by treating samples as replicates

#get results
results.filtered.0.190.38 <- results(dds.filtered.0.190.38, alpha = 0.05)
summary(results.filtered.0.190.38)

Geneids <- rownames(countsTable.filtered.0.190.38)
regulated.genes.filtered <- as.data.frame(results.filtered.0.190.38) %>% mutate(Geneid = Geneids, regulation = ifelse(log2FoldChange > 0, "positive", "negative"))
regulated.genes.filtered <- filter(regulated.genes.filtered, padj < 0.05)

#modify allCounts.filtered to include regulation information
allCounts.filtered <- left_join(allCounts.filtered, regulated.genes.filtered, by = "Geneid")

#modify allCounts.tidy.filterd to include regulation information
allCounts.tidy.filtered <- left_join(allCounts.tidy.filtered, regulated.genes.filtered, by = "Geneid")
```

```{r, echo = F}
#gather nice wide data frame of normalized counts to cluster 
countsToCluster <- allCounts.filtered %>% filter(Geneid %in% regulated.genes.filtered$Geneid) %>%
   select(norm0.00_A, norm0.00_C, norm0.35_A, norm0.35_C, norm11.59_A, norm11.59_B, norm11.59_C, norm20.40_B, norm20.40_C, norm48.37_A,  norm48.37_B, norm48.37_C, norm129.96_A, norm129.96_B, norm129.96_C, norm190.38_A, norm190.38_B, norm190.38_C)
#save Geneid's as rownames
rownames(countsToCluster) <- filter(allCounts.filtered, Geneid %in% regulated.genes.filtered$Geneid)$Geneid

#generate dissmilarity matrix based on spearman correlation
dissimilarity.matrix <- 1 - cor(t(countsToCluster), method = "spearman")

pam.clusters <- c()
pam.medoids <- c()
  pam <- pam(dissimilarity.matrix, 6)
  thisCluster <- as.data.frame(pam$clustering)
  pam.medoids <- c(pam.medoids, rownames(pam$medoids))
  
pam.clusters <- thisCluster %>%
  mutate(Geneid = rownames(countsToCluster))
pam.medoids <- data.frame(pam.medoids, 6)
  
allCounts.tidy.filtered <- left_join(allCounts.tidy.filtered, pam.clusters) 
deep.regulated.tidy <- filter(allCounts.tidy.filtered, regulation != "NA")
```

```{r}
medoids.k6 <- filter(deep.regulated.tidy, Geneid %in% pam.medoids$pam.medoids)
```

```{r}
#k = 6
#show medoids alone
ggplot(data = medoids.k6, aes(x = `RpoS%`, y = normCountScaled)) + geom_point() + geom_line(data = medoids.k6, aes(y = meanScaled), size = 2.3) + coord_cartesian(ylim = c(0,2)) + theme(legend.position = "none") + facet_grid(. ~ `pam$clustering`) + ggtitle("PAM medoids, k=6")+ ylab("Gene Expression")
```

```{r}
# k = 6, show genes with medoids
ggplot(data = deep.regulated.tidy, aes(x = `RpoS%`, y = normCountScaled, col=Geneid)) + geom_point() + geom_line(data = deep.regulated.tidy, aes(y = meanScaled), linetype = "dashed") + coord_cartesian(ylim = c(0,2.3)) + theme(legend.position = "none") + facet_grid(. ~ `pam$clustering`) + ggtitle("Gene Expression Clustering, PAM k=6; Medoids Overlaid in black") + ylab("Gene Expression") + geom_line(data = medoids.k6, aes(y = meanScaled), col = "black", size = 2) + theme(plot.title = element_text(face = "bold"))
```


In the shallow and deep data, a similar proportion of genes are found to be differentially expressed across the most extreme levels, 0% and 190.38% RpoS. 

In both the shallow and deep data, we notice similar "common shapes" (medoids) and a considerable amount of nonmonoticity in the gene expression across different levels of RpoS. 

## How Similar are the Deep and Shallow Clusterings?

ARI: Deep (0.00_B, 0.35_B, 20.40_A removed) vs. Shallow:

```{r}
medoids.compare <- inner_join(deep.regulated.tidy, shallow.regulated.tidy,
                              by = c("Geneid", "feature", "Geneid1", "Geneid2", "feature1", "feature2", "start.gene", "end.gene", "start.bnum", "end.bnum", "genename", "bnum", "RpoS%", "replicate")) 

round(adjustedRandIndex(medoids.compare$`pam$clustering.x`, medoids.compare$`pam$clustering.y`),3)
```

The adjusted rand index suggests above that the clusterings for $k = 6$ of the deep data (with 0.00_B, 0.35_B, 20.40_A removed) and shallow datasets are more similar than dissimilar, but certainly don't agree wholeheartedly.

ARI: Deep (0.00_B, 0.35_B, 20.40_A removed) vs. Deep (0.00_B, 0.35_B, 20.40_A included)

```{r}
medoids.compare <- inner_join(deep.regulated.tidy, extra.deep.regulated.tidy,
                              by = c("Geneid", "feature", "Geneid1", "Geneid2", "feature1", "feature2", "start.gene", "end.gene", "start.bnum", "end.bnum", "genename", "bnum", "RpoS%", "replicate")) 

round(adjustedRandIndex(medoids.compare$`pam$clustering.x`, medoids.compare$`pam$clustering.y`),3)
```

The clusterings for the deep data with and without 0.00_B, 0.35_B, and 20.40_A are more similar than the deep vs. shallow, but certainly don't agree wholeheartedly either. 