---
title: "Questions Regarding 2017 Data"
author: Madison Hobbs, Scripps College; Jo Hardin, Pomona College Mathematics Department; Dan
  Stoebel, Harvey Mudd College Biology Department
date: "5 July 2017"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE, echo = FALSE}
knitr::opts_chunk$set(cache = TRUE)
library(dplyr)
library(tidyr)
library(ggplot2)
library(DESeq2)
library(GGally)
```
```{r, echo = FALSE}
#load data
CodYcounts <- read.csv("CodY.data.csv") # to compare
# our data
allCounts <- read.csv("DMS2670_LB_NC_000913.tsv", header = T, sep = "\t") 
allCounts <- allCounts %>%
  group_by(Geneid) %>%
  mutate(genename = ifelse(startsWith(as.character(Geneid), "IGR") | startsWith(as.character(Geneid), "AS_IGR"), paste(unlist(strsplit(as.character(Geneid), "[,]"))[3], "-", unlist(strsplit(as.character(Geneid), "[,]"))[7]), unlist(strsplit(as.character(Geneid), "[:]"))[3]),
         code = unlist(strsplit(as.character(Geneid), "[:]"))[1],
         geneLabel = paste(code, genename)) %>%
  ungroup() %>%
  select(Geneid, geneLabel, code, genename, 
         "0.00_A" = A0, "0.00_B" = B0, "0.00_C" = C0, "0.35_A" = A10..5, "0.35_B" = B10..5, "0.35_C" = C10..5, "11.59_A" = A5x10.5, "11.59_B" = B5x10.5, "11.59_C" = C5x10.5, "20.40_A" = A10..4, "20.40_B" = B10..4, "20.40_C" = C10..4, "48.37_A" = A10..3, "48.37_B" = B10..3, "48.37_C" = C10..3, "129.96_A" = A2x10..3, "129.96_B" = B2x10..3, "129.96_C" = C2x10..3, "190.38_A" = A5x10..3, "190.38_B" = B5x10..3, "190.38_C" = C5x10..3) 
# note: did not the 100.00 samples in the table

# tidy data
countsTable.all.tidy <- allCounts %>%
  gather(cond.samp, rawCount, -Geneid, -geneLabel, -code, -genename) %>%
  mutate(cond.samps = cond.samp) %>%
  separate(cond.samp, c("condition", "sample"), sep = "_")
CodYcounts.tidy <- CodYcounts %>%
  gather(cond.samps, rawCount, -gene)
```
#Table of Contents
0. The Big Question
1. Total Counts for *E. coli* and *B. subtilis*
2. Interquartile Range of Read Counts by Sample
3. Highly Expressed Genes in *E. Coli* and *B. subtilis* Data Sets
4. Correlations Between Replicates
5. Proportion of Reads From Each Feature Within Each Sample in the *E. coli* Data Set
6. Additional notes

#0. A question for you

A number of observations about our data that concern us:

1. The number of mapped reads per sample varied from 1.4 x 10^6 to 4.4 x 10^6, perhaps low for what we're trying to do.

2. A single feature, the IGR smpB-intA, has between 26% and 53% of all reads in every sample.

3. As you'll see below, the median number of counts in each sample is less than 15, and 75% of all features have fewer than 50 counts in **all** samples. That majority of samples have substantially lower counts.

4. The (Spearman) correlation between replicates is < 90% for some pairs of replicates, and < 95% for most.

5. The proportion of reads in a sample that map to CDS, IGR, or ncRNA differ drammatically between replicates. 

### Our question:

Do these problems (a few very high count transcripts, most genes with very low counts, and correlations lower than we might like) seem like problems that could be fixed with deeper sequencing of each sample? Or do you think something else is at play here?

Details, including plots, are below.

Whenever possible, we constrast our data set to data from a paper on which you collaborated, "Hierarchical expression of genes controlled by the *Bacillus subtilis* global regulatory protein CodY". We used this data set because we are trying to do something similar (look at expression profiles across conditions) rather than just look at differential expression. 

Unless otherwise specified, *E. coli* data used in this analysis includes all gene features (sens and antisens CDS, IGR, ncRNA, tRNA, rRNA) and will exclude the wild type condition because it does not have the rssB deletion consistent among all other samples.

#1. Total Counts
```{r, echo = F, fig.height=4}
totalCounts.Ecoli <- allCounts %>% select(`0.00_A`, `0.00_B`, `0.00_C`, 
         `0.35_A`, `0.35_B`, `0.35_C`, 
         `11.59_A`, `11.59_B`, `11.59_C`, 
         `20.40_A`, `20.40_B`, `20.40_C`, 
         `48.37_A`, `48.37_B`, `48.37_C`, 
         `129.96_A`, `129.96_B`, `129.96_C`, 
         `190.38_A`, `190.38_B`, `190.38_C`) %>% summarize_each(funs(sum)) %>%
  gather(cond.samps, total.count)
ggplot(data = totalCounts.Ecoli, aes(x = cond.samps, y=total.count)) + geom_bar(stat = "identity") + ylab("Total Count") + theme(axis.text.x = element_text(angle = 60, hjust=1)) + xlab("Sample") + ggtitle("E. coli Total Counts by Sample")
```
```{r, echo = F, fig.height=4}

totalCounts.Bsub <- CodYcounts %>% select(WT1, WT2, F71Y1, F71Y2, R61K1, R61K2, R61H1, R61H2, null1, null2) %>% summarize_each(funs(sum)) %>%
  gather(cond.samps, total.count)
ggplot(data = totalCounts.Bsub, aes(x = cond.samps, y=total.count)) + geom_bar(stat = "identity") + ylab("Total Count") + theme(axis.text.x = element_text(angle = 60, hjust=1)) + xlab("Sample") + ggtitle("B. subtilis Total Counts by Sample")
```

The plots above show the total count (unnormalized) for each sample of the *E. coli* and *B. subtilis* data sets respectively. The *B. subtilis* data set has higher counts across the board, and their range is 6,249,673 and the maximum being around 1.67 times the minimum. The *E. coli* data has lower counts and more variation in its total counts by sample, with a range of 2,925,511 and the maximum being around 3.04 times the minimum.

#2. Medians and Third Quartiles of E. coli and B. subtilis counts
```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4}
# use DESeq2 to get size factors counts across all samples, all gene features
countsTable.all <- allCounts %>%
  select(`0.00_A`, `0.00_B`, `0.00_C`, 
         `0.35_A`, `0.35_B`, `0.35_C`, 
         `11.59_A`, `11.59_B`, `11.59_C`, 
         `20.40_A`, `20.40_B`, `20.40_C`, 
         `48.37_A`, `48.37_B`, `48.37_C`, 
         `129.96_A`, `129.96_B`, `129.96_C`, 
         `190.38_A`, `190.38_B`, `190.38_C`)
# define conditions
condition.all <- as.factor(c("0.00", "0.00", "0.00", "0.35", "0.35", "0.35", "11.59", "11.59", "11.59", "20.40",  "20.40",  "20.40", "48.37", "48.37", "48.37", "129.96", "129.96", "129.96", "190.38", "190.38", "190.38"))
coldata.all <- as.data.frame(row.names =  colnames(countsTable.all), condition.all)
cds.all <- DESeqDataSetFromMatrix(countData = countsTable.all, colData = coldata.all, design = ~condition.all)
cds.all <- DESeq(cds.all)
countsTable.all.normalized <- as.data.frame(counts(cds.all, normalized = TRUE)) %>%
  mutate(Geneid = allCounts$Geneid, geneLabel = allCounts$geneLabel, code = allCounts$code, genename = allCounts$genename)
# tidy countsTable normalized data
countsTable.all.normalized.tidy <-  countsTable.all.normalized %>%
  gather(cond.samp, normCount, -Geneid, -geneLabel, -code, -genename) %>%
  mutate(cond.samps = cond.samp) %>%
  separate(cond.samp, c("condition", "sample"), sep = "_") %>%
  mutate(condition = as.numeric(condition))

#extract size factors
allWaySizeFactors <- as.data.frame(sizeFactors(cds.all)) 
cond.samp = rownames(allWaySizeFactors)
allWaySizeFactors <- allWaySizeFactors %>%
  mutate(cond.samp = cond.samp)
seqDepth <- as.data.frame(colSums(countsTable.all)) %>% mutate (cond.samp = cond.samp)

seqDepth.sizeFactors <- inner_join(allWaySizeFactors, seqDepth, by = "cond.samp") %>%
  mutate(sizeFactors.allRegions = sizeFactors(cds.all), 
         sequencingDepth.allRegions = colSums(countsTable.all)) %>%
  select(cond.samp, sizeFactors.allRegions, sequencingDepth.allRegions) %>%
  mutate(cond.samps = cond.samp) %>%
  separate(cond.samp, c("condition", "sample"), sep = "_") %>%
  mutate(condition = as.numeric(condition))

summary.all <- countsTable.all.tidy %>% group_by(cond.samps) %>% summarise(min = min(rawCount), Q1 = quantile(rawCount, .25), median = median(rawCount), Q3 = quantile(rawCount, .75), max = max(rawCount)) %>%
  filter(cond.samps %in% c("100.00_A", "100.00_B", "100.00_C") == FALSE)
summary.all <- inner_join(summary.all, seqDepth.sizeFactors, by = "cond.samps") %>% arrange(as.numeric(condition))
summary.all
```

```{r}
#codY data DESeq2
countsTable.CodY <- CodYcounts %>%
  select(WT1, WT2, F71Y1, F71Y2, R61K1, R61K2, R61H1, R61H2, null1, null2)
# define conditions
condition.CodY <- as.factor(c("WT", "WT", "F71Y", "F71Y", "R61K", "R61K", "R61H", "R61H", "null", "null"))
coldata.CodY <- as.data.frame(row.names =  colnames(countsTable.CodY), condition.CodY)
dds.CodY <- DESeqDataSetFromMatrix(countData = countsTable.CodY, colData = coldata.CodY, design = ~condition.CodY)
dds.CodY <- DESeq(dds.CodY)

summary.CodY <- as.data.frame(sizeFactors(dds.CodY)) 
cond.samps <- rownames(summary.CodY)
summary.CodY <- summary.CodY %>% mutate(cond.samps = cond.samps, size.factor = sizeFactors(dds.CodY), condition = c("WT", "WT", "F71Y", "F71Y", "R61K", "R61K", "R61H", "R61H", "null", "null"))
  
```
```{r, echo=F, fig.height=4}
ggplot(data = countsTable.all.tidy, aes(x = cond.samps, y = rawCount)) + geom_boxplot(outlier.color = "NA", coef = 0) + coord_cartesian(ylim = c(0,60)) + theme(axis.text.x = element_text(angle = 60, hjust=1)) + xlab("Condition_Sample") + ylab("Read Count") + ggtitle("E. coli Read Count Interquartile Range")
```
```{r, echo = FALSE, fig.height=4}
ggplot(data = CodYcounts.tidy, aes(x = cond.samps, y = rawCount)) + geom_boxplot(outlier.color = "NA", coef = 0) + theme(axis.text.x = element_text(angle = 60, hjust=1)) + coord_cartesian(ylim = c(0, 3000)) + ggtitle("B. subtilis Read Counts Interquartile Range") + xlab("Condition_Sample") + ylab("Read Count")
```

In the above boxplots, the x-axis represents each of the samples, and the y-axis represents the read count (unnormalized) for each sample. The read counts for the *E. coli* data display much more variability across the different samples than we see in the *B. subtilis* data; the medians and thrid quartiles both vary greatly across different samples (see Note 1). The *B. subtilis* data set has median read counts of around 400 for each sample, and third quartiles in the thousands. The vast majority of genes in *E. coli* data set, on the other hand, have read counts below 40 for each sample, with third quartiles ranging from 3 to 51 and sample median read counts between 1 and 11.

The large differences in medians result in large difference in the size factors that DEseq uses to normalize the data, see Note 2.

Thus, recalling the total counts from above, in the *E. coli* data set there exist huge differences between the third quartiles and the maximum of each sample. We find that a few genes dominate the read counts for each sample in the *E. coli* data. (If you are intersted, you can see the full boxplots in Note 2.)

#3. Highly Expressed Genes in E. coli and B. subtilis Data Sets
```{r, echo =F}
sampleMaxes.Ecoli <- countsTable.all.tidy %>% group_by(cond.samps) %>%
  summarise(maxCount = max(rawCount)) 
MaxGenes.Ecoli <- countsTable.all.tidy %>% 
  filter(rawCount %in% sampleMaxes.Ecoli$maxCount) %>% select(Geneid, cond.samps)
sampleMaxes.Ecoli <- inner_join(sampleMaxes.Ecoli, MaxGenes.Ecoli, by = "cond.samps")
sampleMaxes.Ecoli <- inner_join(sampleMaxes.Ecoli, totalCounts.Ecoli, by = "cond.samps") %>% mutate(propTotalCount = maxCount/total.count)
 
sampleMaxes.Bsub <- CodYcounts.tidy %>% group_by(cond.samps) %>%
  summarise(maxCount = max(rawCount))
MaxGenes.Bsub <- CodYcounts.tidy %>% filter(rawCount %in% sampleMaxes.Bsub$maxCount) %>%
  select(gene, cond.samps)
sampleMaxes.Bsub <- inner_join(sampleMaxes.Bsub, MaxGenes.Bsub, by = "cond.samps")
sampleMaxes.Bsub <- inner_join(sampleMaxes.Bsub, totalCounts.Bsub, by = "cond.samps") %>% mutate(propTotalCount = maxCount/total.count)
```

In the *E. coli* data set, IGR smpB-intA has the highest read count in every sample, with between 708,292 and 1,597,980 counts per sample. IGR smpB-intA alone accounts for between 26% and 53% of the total reads for each sample. Within the set of IGR reads, IGR smpB-intA's reads alone constitute 85% to 98% of all sample-wise reads.

In the *B. subtilis* data set, there is also one gene which has the highest read count in every sample. This gene is BSU39230_wapA with between 243,705 and 519,330 counts per sample. BSU39230_wapA accounts for only between 2% and 4% of the total reads for each sample. 

IGR smpB-intA is not the only gene with a high read count in the *E. coli* data, as we recall from the boxplots above. There are six genes whose counts exceed 100,000 and dominate the samples. These are CDS cspE, CDS rmf, IGR smpB-intA, ncRNA rna106, ncRNA rna 69, and ncRNA rna98. Defining our RpoS regulon as genes which are differentially expressed between the lowest and highest conditions, 0% (knockout) and 190.38%, all but CDS rmf are found in the regulon we would be considering for profile analysis (similar to the clustering analysis done in "Hierarchical expression of genes controlled by the Bacillus subtilis global regulatory protein CodY").

#4. Correlations Between Replicates
```{r, echo=FALSE}
A_B <- cor(allCounts$`0.00_A`, allCounts$`0.00_B`)
B_C <- cor(allCounts$`0.00_B`, allCounts$`0.00_C`)
C_A <- cor(allCounts$`0.00_C`, allCounts$`0.00_A`)
ABC.corr0.00 <- as.data.frame(cbind(A_B, B_C, C_A))

A_B <- cor(allCounts$`0.35_A`, allCounts$`0.35_B`)
B_C <- cor(allCounts$`0.35_B`, allCounts$`0.35_C`)
C_A <- cor(allCounts$`0.35_C`, allCounts$`0.35_A`)
ABC.corr0.35 <- as.data.frame(cbind(A_B, B_C, C_A))

A_B <- cor(allCounts$`11.59_A`, allCounts$`11.59_B`)
B_C <- cor(allCounts$`11.59_B`, allCounts$`11.59_C`)
C_A <- cor(allCounts$`11.59_C`, allCounts$`11.59_A`)
ABC.corr11.59 <- as.data.frame(cbind(A_B, B_C, C_A))

A_B <- cor(allCounts$`20.40_A`, allCounts$`20.40_B`)
B_C <- cor(allCounts$`20.40_B`, allCounts$`20.40_C`)
C_A <- cor(allCounts$`20.40_C`, allCounts$`20.40_A`)
ABC.corr20.40 <- as.data.frame(cbind(A_B, B_C, C_A))

A_B <- cor(allCounts$`48.37_A`, allCounts$`48.37_B`)
B_C <- cor(allCounts$`48.37_B`, allCounts$`48.37_C`)
C_A <- cor(allCounts$`48.37_C`, allCounts$`48.37_A`)
ABC.corr48.37 <- as.data.frame(cbind(A_B, B_C, C_A))

A_B <- cor(allCounts$`129.96_A`, allCounts$`129.96_B`)
B_C <- cor(allCounts$`129.96_B`, allCounts$`129.96_C`)
C_A <- cor(allCounts$`129.96_C`, allCounts$`129.96_A`)
ABC.corr129.96 <- as.data.frame(cbind(A_B, B_C, C_A))

A_B <- cor(allCounts$`190.38_A`, allCounts$`190.38_B`)
B_C <- cor(allCounts$`190.38_B`, allCounts$`190.38_C`)
C_A <- cor(allCounts$`190.38_C`, allCounts$`190.38_A`)
ABC.corr190.38 <- as.data.frame(cbind(A_B, B_C, C_A))

replicate.correlations <- as.data.frame(rbind(ABC.corr0.00, ABC.corr0.35, ABC.corr11.59, ABC.corr20.40, ABC.corr48.37, ABC.corr129.96, ABC.corr190.38))
rownames(replicate.correlations) = c("0.00", "0.35", "11.59", "20.40", "48.37", "129.96", "190.38")

A_B <- cor(allCounts$`0.00_A`, allCounts$`0.00_B`, method = "spearman")
B_C <- cor(allCounts$`0.00_B`, allCounts$`0.00_C`, method = "spearman")
C_A <- cor(allCounts$`0.00_C`, allCounts$`0.00_A`, method = "spearman")
ABC.corr0.00 <- as.data.frame(cbind(A_B, B_C, C_A))

A_B <- cor(allCounts$`0.35_A`, allCounts$`0.35_B`, method = "spearman")
B_C <- cor(allCounts$`0.35_B`, allCounts$`0.35_C`, method = "spearman")
C_A <- cor(allCounts$`0.35_C`, allCounts$`0.35_A`, method = "spearman")
ABC.corr0.35 <- as.data.frame(cbind(A_B, B_C, C_A))

A_B <- cor(allCounts$`11.59_A`, allCounts$`11.59_B`, method = "spearman")
B_C <- cor(allCounts$`11.59_B`, allCounts$`11.59_C`, method = "spearman")
C_A <- cor(allCounts$`11.59_C`, allCounts$`11.59_A`, method = "spearman")
ABC.corr11.59 <- as.data.frame(cbind(A_B, B_C, C_A))

A_B <- cor(allCounts$`20.40_A`, allCounts$`20.40_B`, method = "spearman")
B_C <- cor(allCounts$`20.40_B`, allCounts$`20.40_C`, method = "spearman")
C_A <- cor(allCounts$`20.40_C`, allCounts$`20.40_A`, method = "spearman")
ABC.corr20.40 <- as.data.frame(cbind(A_B, B_C, C_A))

A_B <- cor(allCounts$`48.37_A`, allCounts$`48.37_B`, method = "spearman")
B_C <- cor(allCounts$`48.37_B`, allCounts$`48.37_C`, method = "spearman")
C_A <- cor(allCounts$`48.37_C`, allCounts$`48.37_A`, method = "spearman")
ABC.corr48.37 <- as.data.frame(cbind(A_B, B_C, C_A))

A_B <- cor(allCounts$`129.96_A`, allCounts$`129.96_B`, method = "spearman")
B_C <- cor(allCounts$`129.96_B`, allCounts$`129.96_C`, method = "spearman")
C_A <- cor(allCounts$`129.96_C`, allCounts$`129.96_A`, method = "spearman")
ABC.corr129.96 <- as.data.frame(cbind(A_B, B_C, C_A))

A_B <- cor(allCounts$`190.38_A`, allCounts$`190.38_B`, method = "spearman")
B_C <- cor(allCounts$`190.38_B`, allCounts$`190.38_C`, method = "spearman")
C_A <- cor(allCounts$`190.38_C`, allCounts$`190.38_A`, method = "spearman")
ABC.corr190.38 <- as.data.frame(cbind(A_B, B_C, C_A))

replicate.correlations.spearman <- as.data.frame(rbind(ABC.corr0.00, ABC.corr0.35, ABC.corr11.59, ABC.corr20.40, ABC.corr48.37, ABC.corr129.96, ABC.corr190.38)) %>%
  mutate(condition = c("0.00", "0.35", "11.59", "20.40", "48.37", "129.96", "190.38")) %>%
  gather(sample.comparison, replicate.correlation, -condition)
```
```{r, echo=F}
CodYcounts.tidy <- CodYcounts.tidy %>%
  mutate(cond.samp = cond.samps) %>%
  separate(cond.samp, c("condition", "sample"), sep = -2)

WT <- cor(CodYcounts$WT1, CodYcounts$WT2, method = "spearman")
F71Y <- cor(CodYcounts$F71Y1, CodYcounts$F71Y2, method ="spearman")
R61K <- cor(CodYcounts$R61K1, CodYcounts$R61K2, method = "spearman")
R61H <- cor(CodYcounts$R61H1, CodYcounts$R61H2, method = "spearman")
null <- cor(CodYcounts$null1, CodYcounts$null2, method = "spearman")

replicate.correlations.spearman.CodY <- data.frame(WT, F71Y, R61K, R61H, null)
```
```{r, echo=FALSE, fig.height=4}
ggplot(data=replicate.correlations.spearman, aes(x = condition, y = replicate.correlation)) + geom_point(size = 2, alpha = .7) +  xlab("Condition") + ylab("Spearman Correlation") + ggtitle("E. coli Pairwise Replicate Correlations by Condition")
```

Each point on the above plot represents the pairwise Spearman correlation between two of three replicates for each condition. The Spearman correlations range between 0.791 and 0.957. For comparison, Spearman correlations between the two replicates at each condition for B. subtilis were between 0.973 and 0.992. Some of the Pearson correlation coefficients are shown in Note 4.

#5. Proportion of Reads From Each Feature Within Each Sample in the E. coli Data Set
```{r, echo=FALSE, fig.height=4}
sample.totals <- countsTable.all.tidy %>% 
  group_by(condition, sample, code) %>% summarize(countByRegion = sum(rawCount)) %>% ungroup() %>% 
  group_by(condition, sample) %>% mutate(sampleTotal = sum(countByRegion),
                                  regionProportion = countByRegion/sampleTotal)

ggplot(data = sample.totals, aes(x = code, y = regionProportion, col = condition)) + geom_point() + xlab("Gene Feature") + ylab("Proportion of Sample's Total Raw Counts") + ggtitle("E. coli Proportion of Reads From Each Feature in Each Sample")
```

In addition to the perhaps poor coorelation between samples, we see that the proportion of each sample that is annotated as CDS, IGR, and ncRNA differs dramatticaly between replicates. The proportions in the above plot are calculated in the following way. Summing each sample's raw read counts, we obtain the total raw read count for each sample. Then, within each sample, we group by gene feature to obtain a total raw read count for each feature within each sample. We then take the latter over the total raw read count for each sample to calculate the proportion of reads for each feature within each sample. 
As evidenced by the above plot, the proportion of per-sample gene feature reads is not consistent within each gene feature. CDS is perhaps the most striking gene feature, with proportions varying from 0.059 to 0.473 depending on the sample. IGR and ncRNA fare better, but still show a difference of approximately 0.2 and 0.3, respectively, between their lowest and highest proportions.

Interestingly, for the samples with single digit percentages of CDS, the bulk of those samples are comprised of IGR and ncRNA. For example, 5.9% of the reads in sample 0.00_B were CDS, while 50.6% of its reads were IGR and 39.9% of its reads were ncRNA.

#6. Additional notes

##Note 1: 

We understand that the summed sens and antisens CDS counts are a more appropriate direct comparison to the Bacillus subtilis data. When creating box plots of the raw counts by sample for just this portion of the data (below), we notice that the counts are not as high as within the entire data set (no raw count exceeds 200,000). We also notice that the medians and thrid quartiles are higher, so the lower 75% of the AS_CDS+CDS data has higher read counts than the lower 75% of all gene features, but the lower 75% within AS_CDS+CDS still does not exceed 200 for nearly all samples. There still exist substantial differences between the median and third quartile read counts of each sample, appearing to vary even more greatly than in the combined data.

```{r, echo = FALSE, fig.height=4}
allCounts.toCompare <- allCounts %>% filter(code %in% c("AS_CDS", "CDS")) %>%
  select(-(1:2)) # just select code name, genename, and all counts
# sum the AS_CDS and CDS rows
allCounts.toCompare <- aggregate(allCounts.toCompare[3:23], by = list(genename = allCounts.toCompare$genename), FUN = sum)
#make tidy
toCompare.tidy <- allCounts.toCompare %>%
  gather(cond.samps, rawCount, -genename)
# make plots
ggplot(data = toCompare.tidy, aes(x = cond.samps, y = rawCount)) + geom_boxplot() + ggtitle("Summed Sens and Antisens CDS Counts by Sample") + xlab("Condition_Sample") + ylab("CDS + AS_CDS Raw Count")
```
```{r, echo=FALSE, fig.height=4}
ggplot(data = toCompare.tidy, aes(x = cond.samps, y = rawCount)) + geom_boxplot() + coord_cartesian(ylim = c(0, 250)) + ggtitle("Zoomed In Summed Sens and Antisens CDS Counts by Sample") + xlab("Condition_Sample") + ylab("Raw Count")
```


##Note 2:
```{r, echo = FALSE, fig.height=3}
ggplot(data = summary.all, aes(x = condition, y = sizeFactors.allRegions)) + geom_point(alpha = 0.7, col="blue") + xlab("Condition") + ylab("Size Factor") + ggtitle("E. coli Size Factors") + geom_hline(yintercept = 1)
```
```{r, echo = FALSE, fig.height=3}
ggplot(data = summary.CodY, aes(x = condition, y = size.factor)) + geom_point(alpha = 0.7, col = "blue") + geom_hline(yintercept = 1) + xlab("Condition") + ylab("Size Factor") + ggtitle("B. subtilis Size Factors")
```
In the above plots, the size factors for each sample produced by DESeq2 are shown for *E. coli* and for *B. subtilis*. On the x-axis we see the conditions and on the y-axis we have the size factor; each condition having a size factor in blue for each of its replicates (3 replicates in the *E. coli* data and two replicates in the *B. subtilis* data). The size factors produced for the *E. coli* data are much more spread, ranging from 0.13 to 2.08, than the size factors produced for the *B. subtilis* data. This phenomenon makes sense recalling the zoomed-in read count boxplots from part 2a.

##Note 3:

```{r, echo = FALSE, fig.height=3}
ggplot(data = countsTable.all.tidy, aes(x = cond.samps, y = rawCount)) + geom_boxplot() + xlab("Condition_Sample") + ylab("Raw Count") + ggtitle("E. Coli Read Counts by Sample")
```
```{r, echo=FALSE, fig.height=3}
ggplot(data = CodYcounts.tidy, aes(x = cond.samps, y = rawCount)) + geom_boxplot() + ggtitle("B. subtilis Counts by Sample") + xlab("Condition_Sample") + ylab("Raw Count")
```
Looking at the boxplots of *E. coli* and *B. subtilis* read counts by sample with unrestrained y-axis, we see that they both have a number of genes with read counts higher than sample third quartiles. We notice with *E. coli*, however, that the spread between the third quartiles and the highly expressed genes in each sample is much larger. 

In the *E. coli* data, at sample 48.37_C there is one gene at around 200,000 raw counts and the next largest gene is at over 1,000,000 raw counts; at least a five-fold increase. This is not an atypical distance between consecutive reads in *E. coli*, nor is it the largest distance. At the same time, the third quartile for 48.37_C is 37, which is a typical, if not relatively high, third quartile value within the rest of the samples. With the vast majority of genes having counts below 40 while some genes dominate the sample at 1,000,000 reads, we are concerned about what effects this phenomenon has on downstream analysis.

For comparison, the largest gap between any two consecutive, within sample, raw counts in the *B. subtilis* data set is only two-fold (see below).


This two-fold difference is seen in R61H2 between the genes BSU01120_fusA_elongation_factor_G and BSU39230_wapA_cell_wall-associated_protein.

##Note 4: 

Inter-Replicate Correlations by Condition

The Pearson intra-replicate correlations by sample range between 0.9742 to 0.9976, likely influenced by the high count genes. Below are pairs plots to show the most spread correlations (knockout condition) and a more typical distributions of correlations (129.96 condition). 

```{r, echo = FALSE, fig.width=3, fig.height = 4}
allCounts.toPlot <- allCounts
colnames(allCounts.toPlot) <- c("Geneid", "geneLabel", "code", "genename", "A0.00", "B0.00", "C0.00", "A0.35", "B0.35", "C0.35", "A11.59", "B11.59", "C11.59", "A20.40", "B20.40", "C20.40", "A48.37", "B48.37", "C48.37", "A129.96", "B129.96", "C129.96", "A190.38", "B190.38", "C190.38") 
ggpairs(data = allCounts.toPlot, columns = 5:7, title = "0.00", upper = list(continuous = wrap('cor', method = "spearman")))
ggpairs(data = allCounts.toPlot, columns = 20:22, title = "129.96", upper = list(continuous = wrap('cor', method = "spearman")))
```


#EXTRA:

For comparision, we looked at the raw counts in "Hierarchical expression of genes controlled by the Bacillus subtilis global regulatory protein CodY" (footnote 1). Below, we see boxplots of the Bacillus subtilis raw counts by sample, first with unconstrained y-axis, then with a constrained y-axis from 0 to 3,000. As evidenced from the plots, the highest raw counts do not exceed 520,000 and 50% of the counts for each sample are between roughly 60 and 2,500. Thus although the sequencing depth in Bacillus subtilis is lower than the E. coli data, the medians and thrid quartiles of raw counts are higher. Furthermore, there is not such a drastic difference between the medians and third quartiles across samples as we see in the E. coli data. Finally, it is worth noting that the largest gap between any two consecutive, within sample, raw counts is only two-fold (footnote 2); much smaller than in the E. coli data.


Certainly, these differences between the E. coli and Bacillus subtilis datasets could be an artifact of biology, and different roles of genes in their respective genomes. Yet it is still worth wondering why the E. coli raw counts show such a majority of low raw counts in each sample, with a minority of very high counts. 

In the E. coli data, the great difference between third quartiles and maximum read counts per sample does not change appreciably across gene features, and we continue to see large gaps between raw counts by gene whether or not we look at CDS, IGR, ncRNA, etc. individually. We also notice that, generally, genes with large read counts are large across all the samples. 

In the E. coli data, there are 

IGR smpB-intA, mentioned above, has nearly all samples with read counts around 1,000,000. This gene's reads alone constitutes 85% - 98% of all sample-wise IGR reads, and 28% - 53% of reads across all gene features. 

The gene ncRNA rna69, whose total read count per sample ranges from around 300,000 to 800,000, makes up 11% - 75% (mean 50%) of all ncRNA reads, and 1% to 26% (mean 12%) of reads across all gene features. 

The CDS genes, cspE and rmf, make up lower proportions of their own features counts and the counts across all gene features. The raw reads of rmf account for 8% to 24% of sample-wise CDS total reads, and 1% to 7% of reads across all gene features. The raw reads of cspE account for 1% to 8% of sample-wise CDS total counts, and 0.4% to 2% of reads across all gene features. 
```{r, echo=FALSE}
# tables for high count genes with:
# cond.samp  IGR smpB-intA total count  /total IGR count  /total all count
totalCount.IGR <- countsTable.all.tidy %>% filter(code == "IGR") %>% group_by(cond.samps)  %>% summarise(sumIGR = sum(rawCount))
totalCount.all <- countsTable.all.tidy %>% group_by(cond.samps) %>% summarise(sumALL = sum(rawCount))
totals <- inner_join(totalCount.IGR, totalCount.all, by = "cond.samps")
smpB_intA <- inner_join(totals, filter(countsTable.all.tidy, rawCount > 100000, code == "IGR"), by = "cond.samps") %>%
  mutate(propIGR.total = rawCount/sumIGR, propALL.total = rawCount/sumALL) %>%
  select(Geneid, cond.samps, rawCount, propIGR.total, propALL.total)

totalCount.ncRNA <- countsTable.all.tidy %>% filter(code == "ncRNA") %>% group_by(cond.samps) %>%
  summarise(sum.ncRNA = sum(rawCount))
totals <- inner_join(totalCount.ncRNA, totalCount.all, by = "cond.samps")
rna69 <- inner_join(totals, filter(countsTable.all.tidy, genename == "rna69", code == "ncRNA"), by = "cond.samps") %>%
  mutate(prop.ncRNAtotal = rawCount/sum.ncRNA, propALL.total = rawCount/sumALL) %>%
  select(Geneid, cond.samps, rawCount, prop.ncRNAtotal, propALL.total)

totalCount.CDS <- countsTable.all.tidy %>% filter(code == "CDS") %>% group_by(cond.samps) %>%
  summarise(sumCDS = sum(rawCount))
totals <- inner_join(totalCount.CDS, totalCount.all, by  = "cond.samps")
rmf <- inner_join(totals, filter(countsTable.all.tidy, genename == "rmf", code == "CDS"), by = "cond.samps") %>%
  mutate(propCDStotal = rawCount/sumCDS, propALL.total = rawCount/sumALL) %>%
  select(Geneid, cond.samps, rawCount, propCDStotal, propALL.total)
cspE <- inner_join(totals, filter(countsTable.all.tidy, genename == "cspE", code == "CDS"), by = "cond.samps") %>%
  mutate(propCDStotal = rawCount/sumCDS, propALL.total = rawCount/sumALL) %>%
  select(Geneid, cond.samps, rawCount, propCDStotal, propALL.total)
```
Inter-Replicate Correlations by Condition

The Pearson intra-replicate correlations by sample range between 0.9742 to 0.9976, likely influenced by the high count genes. Below are pairs plots to show the most spread correlations (knockout condition) and a more typical distributions of correlations (129.96 condition). 
```{r, echo = FALSE, fig.width=3, fig.height = 4}
allCounts.toPlot <- allCounts
colnames(allCounts.toPlot) <- c("Geneid", "geneLabel", "code", "genename", "A0.00", "B0.00", "C0.00", "A0.35", "B0.35", "C0.35", "A11.59", "B11.59", "C11.59", "A20.40", "B20.40", "C20.40", "A48.37", "B48.37", "C48.37", "A129.96", "B129.96", "C129.96", "A190.38", "B190.38", "C190.38") 
ggpairs(data = allCounts.toPlot, columns = 5:7, title = "0.00", upper = list(continuous = wrap('cor', method = "spearman")))
ggpairs(data = allCounts.toPlot, columns = 20:22, title = "129.96", upper = list(continuous = wrap('cor', method = "spearman")))
```

