---
title: "Sequencing Depth (work of MH)"
author: "Jo Hardin"
date: "6/26/2018"
output: pdf_document
---


```{r, warning=FALSE, echo=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      fig.width = 30, fig.height = 10, dev = "jpeg", 
                      fig.align = 'center', fig.keep="all", fig.path = "UCLApics/", 
                      dpi = 500, 
                      cache = TRUE, message = FALSE, warning = FALSE) 
library(tidyverse)
library(stringr)
library(cluster)
library("DESeq2")
library(amap)
library(mclust)

source("rpoSnorm.R")
source("rpoScount.R")
source("rpoSclust.R")
source("rpoSDEclust.R")
source("rpoScorr.R")
source("rpoSDE.R")

sessionInfo()
```


####  Pulling in data

```{r, warning=FALSE, echo=FALSE, message=FALSE}

norm.samps <- c("0.00_A", "0.00_C", "0.35_A", "0.35_C", 
                "11.59_A", "11.59_B", "11.59_C", "20.40_A", "20.40_B",  
                "48.37_A", "48.37_B", "48.37_C", "129.96_A", "129.96_B", "129.96_C", 
                "190.38_A", "190.38_B", "190.38_C")

norm.cond <- as.character(c("0.00", "0.00", "0.35", "0.35", "11.59", "11.59", "11.59", 
                          "20.40", "20.40", "48.37", "48.37", "48.37", "129.96", "129.96", 
                          "129.96", "190.38", "190.38", "190.38"))
#norm.cond <- as.character(c("0.00", "0.00", "0.00", "0.00", "11.59", "11.59", "11.59", 
#                          "20.40", "20.40", "48.37", "48.37", "48.37", "160", "160", 
#                          "160", "160", "160", "160"))

sh.name <- c("DMS2670_LB_NC_000913.tsv")
sh.parse <- countfunc(sh.name, norm.samps, norm.cond, countcutoff = 0)
sh.count <- sh.parse$counts
sh.countT <- sh.parse$countsT

sh.tot <- sh.count %>% 
  dplyr::select(norm.samps) %>%
  dplyr::summarize_all(funs(sum)) %>%
  tidyr::gather(cond.samps, total.count)

ggplot(data = sh.tot, aes(x = cond.samps, y=total.count)) + geom_bar(stat = "identity") + ylab("Total Count") + theme(axis.text.x = element_text(angle = 60, hjust=1)) + xlab("Sample") + ggtitle("E. coli Total SHALLOW Counts by Sample")

dp.name <- c("DMS2670_LB_NC_000913_deeper.tsv")
dp.parse <- countfunc(dp.name, norm.samps, norm.cond, countcutoff = 0)
dp.count <- dp.parse$counts
dp.countT <- dp.parse$countsT

dp.tot <- dp.count %>% 
  dplyr::select(norm.samps) %>%
  dplyr::summarize_all(funs(sum)) %>%
  tidyr::gather(cond.samps, total.count)

ggplot(data = dp.tot, aes(x = cond.samps, y=total.count)) + 
  geom_bar(stat = "identity") + ylab("Total Count") + 
  theme(axis.text.x = element_text(angle = 60, hjust=1)) + 
  xlab("Sample") + ggtitle("E. coli Total DEEP Counts by Sample")
```


\newpage
####  Normalizing data for sequencing depth

```{r, warning=FALSE, echo=FALSE, message=FALSE}

sh.DE <- DEfunc(sh.count, norm.samps, norm.cond)

sh.depth <- sh.count %>%
  dplyr::select(norm.samps) %>%
  colSums() %>%
  cbind(size = sizeFactors(sh.DE),
        cond.samps = norm.samps, condition = norm.cond) %>%
  as.data.frame(stringsAsFactors = FALSE) %>%
  dplyr::mutate(size = round(as.numeric(size),3))
names(sh.depth) <- c("depth", "size", "cond.samps", "condition")



dp.DE <- DEfunc(dp.count, norm.samps, norm.cond)

dp.depth <- dp.count %>%
  dplyr::select(norm.samps) %>%
  colSums() %>%
  cbind(size = sizeFactors(dp.DE),
        cond.samps = norm.samps, condition = readr::parse_number(norm.cond)) %>%
  as.data.frame(stringsAsFactors = FALSE) %>%
  dplyr::mutate(size = round(as.numeric(size),3))
names(dp.depth) <- c("depth", "size", "cond.samps", "condition")



sh.countT %>% 
  group_by(cond.samps) %>% 
  summarise(min = min(rawCount), Q1 = quantile(rawCount, .25), 
            median = median(rawCount), Q3 = quantile(rawCount, .75), 
            max = max(rawCount)) %>%
  inner_join(sh.depth, by = "cond.samps") %>% 
  arrange(as.numeric(condition)) %>% 
  select(condition, Q1, median, Q3, sizeFactor = size, 
         totalCount = depth)

dp.countT %>% 
  group_by(cond.samps) %>% 
  summarise(min = min(rawCount), Q1 = quantile(rawCount, .25), 
            median = median(rawCount), Q3 = quantile(rawCount, .75), 
            max = max(rawCount)) %>%
  inner_join(dp.depth, by = "cond.samps") %>% 
  arrange(as.numeric(condition)) %>% 
  select(condition, Q1, median, Q3, sizeFactor = size, 
         totalCount = depth) 
```



### Shallow Data

```{r, echo=F}
ggplot(data = sh.countT, aes(x = cond.samps, y = rawCount)) + 
  geom_boxplot() + 
  theme(axis.text.x = element_text(angle = 60, hjust=1)) + 
  xlab("Condition_Sample") + ylab("Read Count") + 
  ggtitle("E. coli SHALLOW Read Count")

ggplot(data = sh.countT, aes(x = cond.samps, y = rawCount)) + 
  geom_boxplot(outlier.color = "NA", coef = 0) + 
  coord_cartesian(ylim = c(0,60))  + theme(axis.text.x = element_text(angle = 60, hjust=1)) + 
  xlab("Condition_Sample") + ylab("Read Count") + 
  ggtitle("E. coli SHALLOW Read Count Interquartile Range")
```

### Deep Data

```{r, echo=F}
ggplot(data = dp.countT, aes(x = cond.samps, y = rawCount)) + 
  geom_boxplot() + 
  theme(axis.text.x = element_text(angle = 60, hjust=1)) + 
  xlab("Condition_Sample") + ylab("Read Count") + 
  ggtitle("E. coli DEEP Read Count")

ggplot(data = dp.countT, aes(x = cond.samps, y = rawCount)) + 
  geom_boxplot(outlier.color = "NA", coef = 0) + 
  coord_cartesian(ylim = c(0,300))  + theme(axis.text.x = element_text(angle = 60, hjust=1)) + 
  xlab("Condition_Sample") + ylab("Read Count") + 
  ggtitle("E. coli DEEP Read Count Interquartile Range")
```

Note that the y-axes on both graphs have different scales, but the distribution of read counts per sample looks similar across the two sequencing depths. The samples with majority lower read count genes in the shallow data still display majority lower read count in the deep data. That is, the samples with the lowest Q3 read count in the shallow data are the same as the samples with lowest read count in the deep data: namely, 0.00_B, 0.35_B, and 20.40_A. The rest of the samples in the deep data follow a similar distribution to those in the shallow data. 

This result could suggest that the deep data is confirming what the shallow data saw, but gene by gene plots are needed to see if this is true.



###  Clustering Data

```{r}

norm.samps <- c("0.00_A", "0.00_C", "0.35_A", "0.35_C", 
                "11.59_A", "11.59_B", "11.59_C", "20.40_A", "20.40_B",  
                "48.37_A", "48.37_B", "48.37_C", "129.96_A", "129.96_B", "129.96_C", 
                "190.38_A", "190.38_B", "190.38_C")

norm.cond <- as.character(c("0.00", "0.00", "0.00", "0.00", "11.59", "11.59", "11.59", 
                          "20.40", "20.40", "48.37", "48.37", "48.37", "160", "160", 
                          "160", "160", "160", "160"))

sh.allCounts <- countfunc(filename = sh.name, norm.samps, norm.cond, countcutoff = 5)
sh.clust <- fullDEclust(sh.allCounts$counts, norm.samps, norm.cond, siglevel = 0.001, 
                       nclust = 5)

dp.allCounts <- countfunc(filename = dp.name, norm.samps, norm.cond, countcutoff = 50)
dp.clust <- fullDEclust(dp.allCounts$counts, norm.samps, norm.cond, siglevel = 0.001, 
                       nclust = 5)

```


```{r}
#k = 6
#show medoids alone
ggplot(data = sh.clust$medoids.k, aes(x = RpoS, y = normCountScaled)) + 
  geom_point() + 
  geom_line(data = sh.clust$medoids.k, aes(y = meanScaled), size = 2.3) + 
  coord_cartesian(ylim = c(0,2)) + 
  theme(legend.position = "none") + 
  facet_grid(. ~ cluster) + 
#  ggtitle("PAM medoids, k=6") + 
  ylab("Gene Expression")


# k = 6, show genes with medoids
ggplot(data = sh.clust$countclust, aes(x = RpoS, y = normCountScaled, col=Geneid)) + 
  geom_point() + 
  geom_line(data = sh.clust$countclust, aes(y = meanScaled), linetype = "dashed") + 
  coord_cartesian(ylim = c(0,2.3)) + theme(legend.position = "none") + 
  facet_grid(. ~ cluster) + 
#  ggtitle("Gene Expression Clustering, PAM k=6; Medoids Overlaid in black") + 
  ylab("Gene Expression") + 
  geom_line(data = sh.clust$medoids.k, aes(y = meanScaled), col = "black", size = 2) + 
  theme(plot.title = element_text(face = "bold"))
```





```{r}
#k = 6
#show medoids alone
ggplot(data = dp.clust$medoids.k, aes(x = RpoS, y = normCountScaled)) + 
  geom_point() + 
  geom_line(data = dp.clust$medoids.k, aes(y = meanScaled), size = 2.3) + 
  coord_cartesian(ylim = c(0,2)) + 
  theme(legend.position = "none") + 
  facet_grid(. ~ cluster) + 
#  ggtitle("PAM medoids, k=6") + 
  ylab("Gene Expression")


# k = 6, show genes with medoids
ggplot(data = dp.clust$countclust, aes(x = RpoS, y = normCountScaled, col=Geneid)) + 
  geom_point() + 
  geom_line(data = dp.clust$countclust, aes(y = meanScaled), linetype = "dashed") + 
  coord_cartesian(ylim = c(0,2.3)) + theme(legend.position = "none") + 
  facet_grid(. ~ cluster) + 
#  ggtitle("Gene Expression Clustering, PAM k=6; Medoids Overlaid in black") + 
  ylab("Gene Expression") + 
  geom_line(data = dp.clust$medoids.k, aes(y = meanScaled), col = "black", size = 2) + 
  theme(plot.title = element_text(face = "bold"))
```
